interactive-tx: Add dual-funding flow, using the interactive tx protocol (feature 28/29) (#851)
lightningnetwork/lightning-rfc: 02-peer-protocol.md

  * [Channel](#channel)
    * [Definition of `channel_id`](#definition-of-channel_id)
    * [Interactive Transaction Construction](#interactive-transaction-construction)
      * [Set-Up and Vocabulary](#set-up-and-vocabulary)
      * [Fee Responsibility](#fee-responsibility)
      * [Overview](#overview)
      * [The `tx_add_input` Message](#the-tx_add_input-message)
      * [The `tx_add_output` Message](#the-tx_add_output-message)

>       * [The `tx_remove_input` and `tx_remote_output` Messages](#the-tx_remove_input-and-tx_remove_output-messages)

Reviewer: typo 'remote' here.

      * [The `tx_complete` Message](#the-tx_complete-message)

>       * [The `tx_signatures` Message](#the-tx_signatures-message)

Reviewer 2: It looks like the links for `tx_init_rbf` and `tx_ack_rbf` were lost in the rebase:

```suggestion
      * [The `tx_signatures` Message](#the-tx_signatures-message)
      * [The `tx_init_rbf` Message](#the-tx_init_rbf-message)
      * [The `tx_ack_rbf` Message](#the-tx_ack_rbf-message)
```

Author: Oops. Will fix...

Note that if A does not send a second input, the negotiation will end without
B's contributions.

        +-------+                       +-------+
        |       |--(1)- tx_add_input -->|       |
        |       |<-(2)- tx_complete ----|       |
        |       |--(3)- tx_add_output ->|       |
        |       |<-(4)- tx_complete ----|       |
        |       |--(5)- tx_add_input -->|       |
        |   A   |<-(6)- tx_add_input ---|   B   |
        |       |<-(7)- tx_add_output --|       |
        |       |--(8)- tx_rm_output -->|       |
        |       |<-(9)- tx_complete ----|       |
        |       |--(10) tx_complete --->|       |
        +-------+                       +-------+

### The `tx_add_input` Message

This message contains a transaction input.

1. type: 66 (`tx_add_input`)
2. data:
    * [`channel_id`:`channel_id`]
    * [`u64`:`serial_id`]
    * [`u16`:`prevtx_len`]
    * [`prevtx_len*byte`:`prevtx`]
    * [`u32`:`prevtx_vout`]
    * [`u32`:`sequence`]
    * [`u16`:`redeemscript_len`]
    * [`redeemscript_len*byte`:`script`]

#### Requirements

The sending node:
  - MUST add all sent inputs to the transaction
  - MUST use a unique `serial_id` for each input

>   - MUST NOT re-transmit inputs it has received from the peer

Author: If you send an input/output to a peer and then receive the exact same one back from a different peer, this is taken to mean that there's a circular chain in the set of peers that you're opening a channel with.

It's fine (and expected) to forward inpus/outputs from one peer to another, however they shouldn't send those back to you.

ids.

### `channel_id`, v2

For channels established using the v2 protocol, the `channel_id` is the
SHA256(lesser-revocation-basepoint || greater-revocation-basepoint),
where the lesser and greater is based off the order of the basepoint.
The basepoints are compact DER-encoded public keys.

If the peer's revocation basepoint is unknown (e.g. `open_channel2`),
a temporary `channel_id` should be found by using a zeroed out basepoint for the unknown peer.

#### Rationale
These values must be remembered by both peers for correct operation anyway.
They're known from the first exchange of messages, obviating the need for
a `temporary_channel_id`. Finally, by mixing information from both sides,

> they avoid `channel_id` collisions.

Author: Reusing a revocation basepoint would make your revocation points known ahead of time. This would be a huge security hole,as it would make any commitment tx that used a re-used basepoint unsafe. The prior peer would know the key to the revocation path of any published commitment tx (provided their counter went as high as the published one etc).

> they avoid `channel_id` collisions.

Author: Which is to say, if your counterparty is reusing revocation basepoints for channels, they've got other problems.

        |       |<-(2)- tx_complete ----|       |
        |       |--(3)- tx_add_output ->|       |
        |       |<-(4)- tx_complete ----|       |
        |       |--(5)- tx_add_input -->|       |
        |   A   |<-(6)- tx_add_input ---|   B   |
        |       |<-(7)- tx_add_output --|       |
        |       |--(8)- tx_rm_output -->|       |
        |       |<-(9)- tx_complete ----|       |
        |       |--(10) tx_complete --->|       |
        +-------+                       +-------+

### The `tx_add_input` Message

This message contains a transaction input.

1. type: 66 (`tx_add_input`)
2. data:
    * [`channel_id`:`channel_id`]
    * [`u64`:`serial_id`]
    * [`u16`:`prevtx_len`]
    * [`prevtx_len*byte`:`prevtx`]
    * [`u32`:`prevtx_vout`]
    * [`u32`:`sequence`]
    * [`u16`:`script_sig_len`]
    * [`script_sig_len*byte`:`script_sig`]

#### Requirements

The sending node:
  - MUST add all sent inputs to the transaction
  - MUST use a unique `serial_id` for each input currently added to the
    transaction
  - MUST NOT re-transmit inputs it has received from the peer
  - if is the *initiator*:
    - MUST send even `serial_id`s

>   - if is the *non-initiator*:

Reviewer 3: I think I would uplift *initiator*/*non-initiator* in its own section. Those requirements play out beyond `tx_add_input` ? Also *initiator* MUST contribute at least one input ?

Author: There's no concrete reason the initiator must contribute one input; in the case of the funding transaction this is true and has been added to that particular section of the document. For a closing transaction, however, *one* of the peers must add the funding output as part of the negotiation but there's no reason it necessarily MUST be the initiator (it is sane to require that they do so for that particular protocol however).

Although repeated between `tx_add_input` and `tx_add_output`, it was thought more practical/useful to the implementer to have the complete set of requirements listed explicitly for each message.

    - MUST send odd `serial_id`s

The receiving node:
  - MUST add all received inputs to the transaction
  - MUST fail the negotiation if:
    - the `prevtx` and `prevtx_vout` are identical to a previously added
      (and not removed) input's
    - `prevtx` is not a valid transaction
    - `prevtx_vout` is greater or equal to the number of outputs on `prevtx`
    - the `prevtx_out` input of `prevtx` is not an `OP_0` to `OP_16`
      followed by a single push
    - the `serial_id` is already included in the transaction
    - the `serial_id` has the wrong parity

>     - if has received 4096 `tx_add_input` messages during this negotiation

Reviewer 3: If transaction size is superior to `MAX_STANDARD_TX_WEIGHT` MUST fail ?

Author: Is `MAX_STANDARD_TX_WEIGHT` a bitcoind constant?

        +-------+                       +-------+

### The `tx_add_input` Message

This message contains a transaction input.

1. type: 66 (`tx_add_input`)
2. data:
    * [`channel_id`:`channel_id`]
    * [`u64`:`serial_id`]
    * [`u16`:`prevtx_len`]
    * [`prevtx_len*byte`:`prevtx`]
    * [`u32`:`prevtx_vout`]
    * [`u32`:`sequence`]
    * [`u16`:`script_sig_len`]
    * [`script_sig_len*byte`:`script_sig`]

#### Requirements

The sending node:
  - MUST add all sent inputs to the transaction
  - MUST use a unique `serial_id` for each input currently added to the
    transaction
  - MUST NOT re-transmit inputs it has received from the peer
  - if is the *initiator*:
    - MUST send even `serial_id`s
  - if is the *non-initiator*:
    - MUST send odd `serial_id`s

The receiving node:
  - MUST add all received inputs to the transaction
  - MUST fail the negotiation if:
    - the `prevtx` and `prevtx_vout` are identical to a previously added
      (and not removed) input's
    - `prevtx` is not a valid transaction
    - `prevtx_vout` is greater or equal to the number of outputs on `prevtx`
    - the `prevtx_out` input of `prevtx` is not an `OP_0` to `OP_16`

>       followed by a single push

Author: Good point, will update.

#### Requirements

The nodes:
  - MUST send this message in succession to conclude this protocol

The receiving node:
  - MUST use the negotiated inputs and outputs to construct a transaction
  - MUST fail the negotiation if:
    - the peer's total input satoshis is less than their outputs
    - the peer's paid feerate does not meet or exceed the agreed `feerate`,
      (based on the `minimum fee`).
    - if is the *non-initiator*:
      - the *initiator*'s fees do not cover the `common` fields
    - there are more than 252 inputs
    - there are more than 252 outputs

#### Rationale
To signal the conclusion of exchange of transaction inputs and outputs.

Upon successful exchange of `tx_complete` messages, both nodes
should construct the transaction and proceed to the next portion of the
protocol. For channel establishment v2, exchanging commitment transactions.

For the `minimum fee` calculation see [BOLT #3](03-transactions.md#calculating-fees-for-collaborative-transaction-construction).

The maximum inputs and outputs are capped at 252. This effectively fixes
the byte size of the input and output counts on the transaction to one (1).

### The `tx_signatures` Message

> 

Author: Additional rationale: it strictly orders expected sigs transmission for multi-dual-funded channels.

Consider double dual-funding open with three participants for the following values; B is coordinating.

```
A, 10M  <---- >  B  5M <----> C 1M
```

Order of `tx_sigs` transmission would be strictly as follows:

```
C -> B
B -> A
A -> B
B -> C
```

  - MUST fail the negotiation if:
    - the message contains an empty `witness_stack`
    - the number of `witness_stack`s does not equal the number of inputs
      added by the sending node
    - the `txid` does not match the txid of the transaction
  - MUST fail the channel if:
    - the `witness_stack` weight lowers the effective `feerate`
      below the agreed upon transaction `feerate`
  - SHOULD apply the `witness`es to the transaction and broadcast it
  - MUST reply with their `tx_signatures` if not already transmitted

#### Rationale
`witness` is the data for a witness element in a witness stack.
Witness elements should *not* include their length.

Witness data must be sorted according to the `serial_id` of
the corresponding input.

While the `minimum fee` is calculated and verified at `tx_complete` conclusion,
it is possible for the fee for the exchanged witness data to be underpaid.
It is the responsibility of the sending peer to correctly account for the
required fee, e.g. a multisig witness stack whose weight exceeds 110.
If the fees paid by the peer (inputs - outputs) does not meet or exceed
the pre-established `feerate`, the receiving peer SHOULD immediately

> fail the channel by broadcasting their commitment transaction.

Author: > Are we sure that's the best failure mode ?

The intent is to stop cooperating with the peer, given that they've failed to follow spec. Failing the channel is a simple way to do so.

> fail the channel by broadcasting their commitment transaction.

Author: @t-bast great point! Will update.

### The `tx_add_input` Message

This message contains a transaction input.

1. type: 66 (`tx_add_input`)
2. data:
    * [`channel_id`:`channel_id`]
    * [`u64`:`serial_id`]
    * [`u16`:`prevtx_len`]
    * [`prevtx_len*byte`:`prevtx`]
    * [`u32`:`prevtx_vout`]
    * [`u32`:`sequence`]
    * [`u16`:`script_sig_len`]
    * [`script_sig_len*byte`:`script_sig`]

#### Requirements

The sending node:
  - MUST add all sent inputs to the transaction
  - MUST use a unique `serial_id` for each input currently added to the
    transaction
  - MUST NOT re-transmit inputs it has received from the peer
  - if is the *initiator*:
    - MUST send even `serial_id`s
  - if is the *non-initiator*:
    - MUST send odd `serial_id`s

The receiving node:
  - MUST add all received inputs to the transaction
  - MUST fail the negotiation if:
    - the `prevtx` and `prevtx_vout` are identical to a previously added
      (and not removed) input's
    - `prevtx` is not a valid transaction
    - `prevtx_vout` is greater or equal to the number of outputs on `prevtx`
    - the `scriptPubKey` of the `prevtx_out` output of `prevtx` does not begin with `OP_0` to `OP_16`

>       followed by a single push of 2 to 40 bytes

Reviewer 4: What is the rationale for not allowing p2sh-embedded segwit inputs?

Reviewer 2: We should remove support for non-segwit inputs, as it creates a lot of issues when allowing dust limits below 546 sats (see https://github.com/lightningnetwork/lightning-rfc/pull/894)

Author: Oh I see. This should be updated to be *either* the `scriptPubKey` or the `script_sig`, if provided. This will encapsulate both cases, and permit P2SH-wrapped segwit spends to be eligible.

Author: Ok updated the text to allow for P2SH-wrapped spends.

Regarding moving to native segwit only (P2W*++), this is a reasonable policy for *outputs* I think, but not for spending P2SH-wrapped inputs, as there's no dust considerations for spends (unlike output creations)

>     - the `scriptPubKey` of the `prevtx_out` output of `prevtx` does not begin with `OP_0` to `OP_16`
>       followed by a single push of 2 to 40 bytes

Reviewer 5: While the intention is quite clear, a dumb-dumb reading of this results in any opcode between `0x00` and `0x60`, as well as the ambiguity of what the total script may be

```suggestion
    - the `scriptPubKey` of the `prevtx_out` output of `prevtx` is not exactly a 1-byte push opcode (for the numeric values 0 to 16) followed by a data push between 2 and 40 bytes
```

cribbing heavily from bip141 text. Always complained we never had a short name for this script template :/

4294967294 (0xFFFFFFFE). See BIP125.

`script_sig` is the scriptSig for the input, with length omitted.
The `script_sig` for non-P2SH-wrapped inputs will be an empty byte array.

### The `tx_add_output` Message

This message adds a transaction output.

1. type: 67 (`tx_add_output`)
2. data:
    * [`channel_id`:`channel_id`]
    * [`u64`:`serial_id`]
    * [`u64`:`sats`]
    * [`u16`:`scriptlen`]
    * [`scriptlen*byte`:`script`]

#### Requirements

Either node:
  - MAY omit this message

The sending node:
  - MUST add all sent outputs to the transaction
  - if is the *initiator*:
    - MUST send even `serial_id`s
  - if is the *non-initiator*:
    - MUST send odd `serial_id`s

The receiving node:
  - MUST add the specified output to the transaction
  - MUST accept P2SH, P2WSH, P2WPKH, P2PKH, P2TR `script`s
  - MAY fail the negotiation if `script` is non-standard
  - MUST fail the negotiation if:
    - the `serial_id` is already included in the transaction
    - the `serial_id` has the wrong parity
    - it has received 4096 `tx_add_output` messages during this negotiation
    - the `sats` amount is less than the `dust_limit`

>     - the `sats` amount is greater than 21,0000,0000 (`MAX_MONEY`)

Reviewer 4: I am a bit confused by the number of zeroes here. There are 8 zeroes here; and I suspect you either wanted 6 or 14.
What are the units?
`sats` is in satoshis I think? If you meant "21 milion bitcoins", the numeric literal should be `21 * 10**6 * 10**8`.

or implement stricter checks on standardness.

### The `tx_remove_input` and `tx_remove_output` Messages
These message removes an input from the transaction.

1. type: 68 (`tx_remove_input`)
2. data:
    * [`channel_id`:`channel_id`]
    * [`u64`:`serial_id`]

This message removes an output from the transaction.

1. type: 69 (`tx_remove_output`)
2. data:
    * [`channel_id`:`channel_id`]
    * [`u64`:`serial_id`]

#### Requirements

The sending node:
  - MUST NOT send a `tx_remove` with a `serial_id` it did not add
    to the transaction or has already removed

The receiving node:
  - MUST remove the indicated input or output from the transaction
  - MUST fail the negotiation if:
    - the input or output identified by the `serial_id` was not added by the
      sender
    - the `serial_id` does not correspond to a currently added input (or output)

### The `tx_complete` Message

This message signals the conclusion of a peer's transaction
contributions.

1. type: 70 (`tx_complete`)
2. data:
    * [`channel_id`:`channel_id`]

#### Requirements

The nodes:
  - MUST send this message in succession to conclude this protocol

> The receiving node:

Reviewer 5: If the receiving node still want to add things it can, yes? 

Basically "once a node has both sent and received a tx_complete message for the channel_id ..." so on and so forth?

Author: The sent/received `tx_complete` needs to be sequential for the negotiation to be considered concluded. Consider the following case:

```
A                           B
tx_complete ->
              <-  tx_add_input
tx_add_input ->
              <- tx_complete
```

Although a `tx_complete` has been exchanged by both parties, they weren't sequential therefore the interactive tx construction protocol is considered still in force.

`A` may send another `tx_complete` at this point to end the negotiation; they may also choose to send `tx_add_output` etc and allow `B` to respond again.

Reviewer 5: ok I'm guessing I misunderstood what "succession" meant or missed that line entirely as per the above. it's clear

ids.

### `channel_id`, v2

For channels established using the v2 protocol, the `channel_id` is the
SHA256(lesser-revocation-basepoint || greater-revocation-basepoint),

> where the lesser and greater is based off the order of the basepoint.

Author: Ah yeah, good point (hehe). Since the definition in `01` is for the word "point", it seems that would apply to base*points* as well. I'll remove the reference to the encoding type here.

- The `dust_limit` for the transaction is known.
- The `nLocktime` for the transaction has been negotiated, or is established
  via convention.
- The transaction version is 2.

### Fee Responsibility

The *initiator* is responsible for paying the fees for the following fields,
to be referred to as the `common fields`.

  - version
  - segwit marker + flag
  - input count
  - output count
  - locktime

The rest of the transaction bytes' fees are the responsibility of
the peer who contributed that input or output via `tx_add_input` or
`tx_add_output`, at the agreed upon `feerate`.

### Overview

The *initiator* initiates the interactive transaction construction
protocol with `tx_add_input`. The *non-initiator* responds with any
of `tx_add_input`, `tx_add_output`, `tx_rm_input`, `tx_rm_output`, or
 `tx_complete`. The protocol continues with the synchronous exchange
of interactive transaction protocol messages until both nodes have sent
and received a consecutive `tx_complete`.

Once peers have exchanged consecutive `tx_complete`s, the
interactive transaction construction protocol is considered concluded.
Both peers should construct the transaction and fail the negotiation

> if an error is discovered.

Author: Ah. So 'fail the negotiation' does have an explicit meaning for the implementation that I've got in c-lightning. My bad for not explicitly including it here!

A 'negotiation fail' consists of two things:
- first, sending a 'warning' message with the error
- next, disconnecting with the peer.

The disconnect has the benefit of resetting the state such that the peer can reconnect and try again from scratch.

> if an error is discovered.

Reviewer 2: Thanks, that's more clear.

I still think that a dedicated message (`abort_tx`?) would make sense though as:

- we may want to send non-fatal warnings (without disconnecting) during the interactive-tx sessions for other reasons
- it could be useful to abort without necessarily disconnecting (one of the parties changed their mind)
- it's more future-proof since it gives us a message that we can easily extend in the future to add data that could help automate retries

WDYT?

> if an error is discovered.

Reviewer 2: That's because you're looking at it through the lens of c-lightning which has a single channel per peer ;)
Imagine I already have a channel with you and payments are steadily flowing through it.
In parallel, we start a dual funding flow.
But actually I've run out of utxos (because I'm also splicing in with someone else in parallel) so I want to abort.
I don't want to have to disconnect to tell you that, because that disrupts our other channel that is processing payments.
Does that make sense?

> if an error is discovered.

Author: Ah yes, you present a very persuasive argument here @t-bast! c-lightning will indeed have this same problem as of our next release lol.

Ok, cool I'll reconfigure it to add an `abort_tx` call, which should be equivalent to hanging up (also).

> if an error is discovered.

Author: Actually, question. Does the `error` message not serve the same purpose here? It should result in the same desired behavior signal: this negotiation is fatally flawed and cannot progress forward.

> if an error is discovered.

Author: Or a `warning`? Either of these is useful for signaling "this conversation is over, let's start again".

> if an error is discovered.

Reviewer 2: I don't think we should use `warning` for that, there could be reasons to send informative `warning`s without failing the negotiation. Maybe `error` would work though, but I still think it's slightly cleaner to have an explicit `abort`, as aborting isn't necessarily an error case (and we may want in the future to enrich the `tx_abort` message with specific options / fields).

I'm not sure about that yet, but I'll be able to report back on this once I've made progress on the implementation!

This protocol is expressly designed to allow for parallel, multi-party
sessions to collectively construct a single transaction. This preserves
the ability to open multiple channels in a single transaction. While
`serial_id`s are generally chosen randomly, to maintain consistent transaction
ordering across all peer sessions, it is simplest to invert the bottom-bit of
received `serial_id` before forwarding them to other peers.

Here are a few example exchanges.

#### *initiator* only

A, *initiator* has two inputs and an output (the funding output).
B, the *non-initiator* has nothing to contribute.
n

        +-------+                       +-------+
        |       |--(1)- tx_add_input -->|       |
        |       |<-(2)- tx_complete ----|       |
        |       |--(3)- tx_add_input -->|       |
        |   A   |<-(4)- tx_complete ----|   B   |
        |       |--(5)- tx_add_output ->|       |
        |       |                       |       |
        |       |--(6)- tx_complete --->|       |

>         |       |<-(7)- tx_complete ----|       |

Author: This is a turn-based protocol, I should mention that explicitly. The peer waits until it has received a message (any message) from the peer before replying. When a peer runs out of things to day, it sends `tx_complete` back for its turn.

>         |       |<-(7)- tx_complete ----|       |

Reviewer 2: Ok, this is much clearer then! If it's a turn-based protocol it's easy, there's no risk of race conditions, I like that :+1:

>         |       |<-(7)- tx_complete ----|       |

Reviewer 2: But in that case Bob's `tx_complete` must come before Alice's `tx_complete` in this diagram, doesn't it?

>         |       |<-(7)- tx_complete ----|       |

Reviewer 4: Right... Why did Alice get to send two messages (5 and 6) without waiting for a reply from Bob in between?

>         |       |<-(7)- tx_complete ----|       |

Reviewer 5: > But in that case Bob's tx_complete must come before Alice's tx_complete in this diagram, doesn't it?

Bob is allowed to continue sending any number of messages until he is done now that he knows Alice is done, then sends his `tx_complete`, without waiting turns.

        +-------+                       +-------+

#### *initiator* and *non-initiator*

A the *initiator* contributes 2 inputs and an output that they
then remove.  B, the *non-initiator*, contributes 1 input and an output,
but waits until A adds a second input before contributing.

Note that if A does not send a second input, the negotiation will end without
B's contributions.

        +-------+                       +-------+
        |       |--(1)- tx_add_input -->|       |
        |       |<-(2)- tx_complete ----|       |
        |       |--(3)- tx_add_output ->|       |
        |       |<-(4)- tx_complete ----|       |
        |       |--(5)- tx_add_input -->|       |
        |   A   |<-(6)- tx_add_input ---|   B   |
        |       |<-(7)- tx_add_output --|       |
        |       |--(8)- tx_rm_output -->|       |
        |       |<-(9)- tx_complete ----|       |
        |       |--(10) tx_complete --->|       |
        +-------+                       +-------+

### The `tx_add_input` Message

This message contains a transaction input.

1. type: 66 (`tx_add_input`)
2. data:

>     * [`channel_id`:`channel_id`]

Author: > if this protocol is abstract (and may be used for other things than opening channels)?

Well, let's consider the cases (in lightning, at least) that we'd want to communicate transactions. There are three cases that come to mind:

- Opening a channel
- Renegotiating a channel (funding)
- Closing a channel

As all of these actions revolve around the "channel", it seems accurate to leave that as the generalized identifier.

>     * [`channel_id`:`channel_id`]

Reviewer 2: That seems fair, we can change the name of this parameter later if we ever use this protocol for something else than a channel :+1:

    * [`u64`:`serial_id`]
    * [`u16`:`prevtx_len`]
    * [`prevtx_len*byte`:`prevtx`]
    * [`u32`:`prevtx_vout`]
    * [`u32`:`sequence`]
    * [`u16`:`script_sig_len`]
    * [`script_sig_len*byte`:`script_sig`]

#### Requirements

The sending node:
  - MUST add all sent inputs to the transaction
  - MUST use a unique `serial_id` for each input currently added to the
    transaction
  - MUST NOT re-transmit inputs it has received from the peer
  - if is the *initiator*:
    - MUST send even `serial_id`s
  - if is the *non-initiator*:
    - MUST send odd `serial_id`s

The receiving node:
  - MUST add all received inputs to the transaction
  - MUST fail the negotiation if:
    - the `prevtx` and `prevtx_vout` are identical to a previously added
      (and not removed) input's
    - `prevtx` is not a valid transaction
    - `prevtx_vout` is greater or equal to the number of outputs on `prevtx`

>     - the `prevtx_out` input of `prevtx` is not an `OP_0` to `OP_16`

Author: this is more precise, will update.

      followed by a single push of 2 to 40 bytes
    - the `serial_id` is already included in the transaction
    - the `serial_id` has the wrong parity
    - if has received 4096 `tx_add_input` messages during this negotiation

#### Rationale
Each node must know the set of the transaction inputs. The *non-initiator*
MAY omit this message.

`serial_id` is a randomly chosen number which uniquely identifies this input.
Inputs in the constructed transaction are sorted by `serial_id`.

> `prevtx_tx` is the serialized transaction that contains the output

Author: > doesn't this conflict with this paragraph in the overview?

```
MUST NOT re-transmit inputs it has received from the peer
```

The intention of this statement is to not send inputs it's received from the peer *back to the same peer*. You can forward them onward other places, but repeating them back to the originator is an error.

> `prevtx_tx` is the serialized transaction that contains the output

Reviewer 2: Gotcha, thanks for clarifying!

this input spends. Used to verify that the input is non-malleable.

`prevtx_vout` is the index of the output being spent.

`sequence` is the sequence number of this input.  Must be less than
4294967294 (0xFFFFFFFE). See BIP125.

`script_sig` is the scriptSig for the input, with length omitted.

> The `script_sig` for non-P2SH-wrapped inputs will be an empty byte.

Author: Yes, this should be an empty, 0-len, array.

### The `tx_add_output` Message

This message adds a transaction output.

1. type: 67 (`tx_add_output`)
2. data:
    * [`channel_id`:`channel_id`]
    * [`u64`:`serial_id`]
    * [`u64`:`sats`]
    * [`u16`:`scriptlen`]
    * [`scriptlen*byte`:`script`]

#### Requirements

Either node:
  - MAY omit this message

The sending node:
  - MUST add all sent outputs to the transaction
  - if is the *initiator*:
    - MUST send even `serial_id`s
  - if is the *non-initiator*:
    - MUST send odd `serial_id`s

The receiving node:
  - MUST add the specified output to the transaction

>   - MUST accept P2SH, P2WSH, P2WPKH, P2PKH `script`s

Author: Yes, probably. Does this require anything other than a mention though? :thinking:

2. data:
    * [`channel_id`:`channel_id`]
    * [`sha256`:`txid`]
    * [`u16`:`num_witnesses`]
    * [`num_witnesses*witness_stack`:`witness_stack`]

1. subtype: `witness_stack`
2. data:
    * [`u16`:`num_input_witness`]
    * [`num_input_witness*witness_element`:`witness_element`]

1. subtype: `witness_element`
2. data:
    * [`u16`:`len`]
    * [`len*byte`:`witness`]

#### Requirements
The sending node:
  - if it has the lowest total satoshis contributed, as defined
    by total `tx_add_input` values:
    - MUST transmit their `tx_signatures` before the peer
  - MUST order the `witness_stack`s by the `serial_id` of the input they
    correspond to
  - number of `witness_stack`s MUST equal the number of inputs they added

The receiving node:
  - MUST fail the negotiation if:
    - if they have not sent a `tx_signatures` message and they are the
      peer with the lowest total satoshis contributed, as defined by
      `tx_add_input` values
    - the message contains an empty `witness_stack`
    - the number of `witness_stack`s does not equal the number of inputs
      added by the sending node
    - the `txid` does not match the txid of the transaction

>   - MUST fail the channel if:

Author: Updated to remove refernence to the channel.

1. subtype: `witness_element`
2. data:
    * [`u16`:`len`]
    * [`len*byte`:`witness`]

#### Requirements

The sending node:
  - if it has the lowest total satoshis contributed, as defined
    by total `tx_add_input` values:
    - MUST transmit their `tx_signatures` before the peer
  - MUST order the `witness_stack`s by the `serial_id` of the input they
    correspond to
  - number of `witness_stack`s MUST equal the number of inputs they added

The receiving node:
  - MUST fail the negotiation if:
    - if they have not sent a `tx_signatures` message and they are the
      peer with the lowest total satoshis contributed, as defined by
      `tx_add_input` values
    - the message contains an empty `witness_stack`
    - the number of `witness_stack`s does not equal the number of inputs
      added by the sending node
    - the `txid` does not match the txid of the transaction
  - SHOULD apply the `witness`es to the transaction and broadcast it
  - MUST reply with their `tx_signatures` if not already transmitted

#### Rationale

The peer with the lowest total of inputs amounts must transmit its
`tx_signatures` first. This gives in a strict ordering of transmission

> for any multiparty tx collaboration.

Reviewer 2: This isn't completely true, we don't specify yet what happens when both peers contribute the same amount.
I think we should mandate that in that case, the initiator starts?
WDYT @niftynei?

Author: Agreed, I think that's a good tie-breaker!

Reviewer 5: FYI simplified updates proposal uses the lexicographical-sorted pubkeys as tie-breaker. Only real difference I can see is initiator doesn't generalize to N party... :shrug:

Author: We could also remove this ordering clause and leave it as unspecified? The current implementation just sends the signatures whenever it has them.

I think ordering makes finding the source of a deadlock easier tho, so might be worth adding for that reason.

Reviewer 2: I'd like to keep an ordering, it makes it much easier to fully specify your state machine when implementing.
I don't mind which one we choose, but I'd like one!

> FYI simplified updates proposal uses the lexicographical-sorted pubkeys as tie-breaker. Only real difference I can see is initiator doesn't generalize to N party... :shrug:

Interesting, that's definitely a good option as the interactive-tx protocol should easily generalize to N party transaction building. We can do that on the funding pubkeys, how does that sound @niftynei?

Reviewer 2: > why would a deadlock be expected?

There is an advantage to being the last party to send your sigs, because then you're sure you have a fully signed transaction whenever you send your sigs. Whereas if you send them first, your peer may keep them but never give you their sigs, and you're unable to publish the transaction yourself but still have to remember it until it's double-spent.

Without a specified ordering, there would be an incentive to always send our sigs last, which would lead to a deadlock if both peers apply this strategy.

Author: > why would a deadlock be expected? signatures aggregation can proceed in any order converging on the same set of signatures (a G-set in CRDT parlance)

If one party fails to send signatures, the negotiation fails. Ordering them makes it obvious which party failed to send signatures.

Author: >  We can do that on the funding pubkeys, how does that sound @niftynei?

Works for me, thanks!

pubkey corresponding to the funding output nothing prevents duplicative channel
ids.

### `channel_id`, v2

For channels established using the v2 protocol, the `channel_id` is the
SHA256(lesser-revocation-basepoint || greater-revocation-basepoint),
where the lesser and greater is based off the order of the basepoint.

If the peer's revocation basepoint is unknown (e.g. `open_channel2`
and `accept_channel2`), a temporary `channel_id` should be found by

> using a zeroed out basepoint for the unknown peer.

Reviewer 2: Yes, that's why it's always using the initiator's basepoint. The unknown party is the non-initiator here.

#### Rationale

These values must be remembered by both peers for correct operation anyway.
They're known from the first exchange of messages, obviating the need for
a `temporary_channel_id`. Finally, by mixing information from both sides,
they avoid `channel_id` collisions.

## Interactive Transaction Construction

Interactive transaction construction allows two peers to collaboratively
build a transaction for broadcast.  This protocol is the foundation
for dual-funded channels establishment (v2).

### Set-Up and Vocabulary

There are two parties to a transaction construction: an *initiator*
and a *non-initiator*.
The *initiator* is the peer which initiates the protocol, e.g.
for channel establishment v2 the *initiator* would be the peer which
sends `open_channel2`.

The protocol makes the following assumptions:

> - The `feerate` for the transaction is known.

Reviewer 3: IIUC, the feerate assumed as known here is `funding_feerate_perkw` from `open_channel2`. Could be precised for clarity.

- The `dust_limit` for the transaction is known.
- The `nLocktime` for the transaction is known.

> - The transaction version is 2.

Reviewer 3: I think in the future we might introduce version 3 if we would like interactive transaction to benefit from the advantage of package RBF. This should be achievable by introducing a new TLV record in `open_channel2`.

See https://github.com/bitcoin/bitcoin/pull/25038

### Fee Responsibility

The *initiator* is responsible for paying the fees for the following fields,
to be referred to as the `common fields`.

  - version
  - segwit marker + flag
  - input count
  - output count
  - locktime

The rest of the transaction bytes' fees are the responsibility of
the peer who contributed that input or output via `tx_add_input` or
`tx_add_output`, at the agreed upon `feerate`.

### Overview

The *initiator* initiates the interactive transaction construction
protocol with `tx_add_input`. The *non-initiator* responds with any
of `tx_add_input`, `tx_add_output`, `tx_remove_input`, `tx_remove_output`, or
`tx_complete`. The protocol continues with the synchronous exchange
of interactive transaction protocol messages until both nodes have sent
and received a consecutive `tx_complete`. This is a turn-based protocol.

Once peers have exchanged consecutive `tx_complete`s, the
interactive transaction construction protocol is considered concluded.
Both peers should construct the transaction and fail the negotiation
if an error is discovered.

This protocol is expressly designed to allow for parallel, multi-party

> sessions to collectively construct a single transaction. This preserves

Reviewer 6: wouldn't the strict ordering requirements of who sends their signatures first cause deadlocks?

i imagine 3 nodes (A, B and C) setting up a triangle of channels where every node contributes the same amount of sats.
when tx_sig has to be sent:
one node (A) has the lexicographically lowest public key and must start, now node B expects tx_sig from A, which include signatures for C's inputs, but A cant send it because it doesnt have them. it also wont get them from C because C waits until A sends tx_sig which contains signatures for B's inputs, and so on...

Reviewer 7: Some notes on the subject written down here: https://github.com/lightning/bolts/pull/863/files#r897441047

This PR is meant to add strict ordering of tx_sig: https://github.com/rustyrussell/lightning-rfc/pull/9

It's applicable to splicing and dual not 100% clear where it should go. Putting it all in dual may be the right way to go.

Reviewer 7: This has implications for `tx_abort` as well.

Reviewer 2: This is a good point @antonilol, in multi-party scenarios the ordering requirements can lead to a deadlock.
I think that the simplest way to work around that is to have:

- nodes support receiving remote signatures early (eclair already does that): even if you're supposed to send first, if your remote sends their signatures first, you should gladly accept them
- when a node tries to combine multiple `interactive-tx` sessions, they should send their sigs first, which removes that deadlock

Does that make sense? It's simple enough and preserves the strict ordering for 2-party txs, which is going to be the majority of cases.

Reviewer 6: > * nodes support receiving remote signatures early (eclair already does that): even if you're supposed to send first, if your remote sends their signatures first, you should gladly accept them.

i think the following must be changed then (and rename "strict ordering"):
```
The receiving node [of `tx_signatures`]:
  - MUST fail the negotiation if:
    - if they have not sent a `tx_signatures` message and they are the
      peer with the lowest total satoshis contributed, as defined by
      `tx_add_input` values
```

> * when a node tries to combine multiple `interactive-tx` sessions, they should send **their sigs** first, which removes that deadlock

**their sigs** includes sigs from the other parties.
in a case with A, B and C where A combines interactive tx sessions with B and C, B expected C's sigs from A, (and doesn't know they are from C), and the other way around.

why not make the signature exchange work like the add_input and add_output, (with the same serial id to identify which inputs the sigs are for). this way all parties can send the sigs they can create to their peers (in separate messages or multiple in one, but don't require to send all in one like tx_sig) and relay sigs received from other peers

Reviewer 2: > in a case with A, B and C where A combines interactive tx sessions with B and C, B expected C's sigs from A, (and doesn't know they are from C), and the other way around.

Right, good catch, I was only focusing on the case where A pipes from B to C, but not the other way around as well (so my scenario actually didn't make any sense).

> why not make the signature exchange work like the add_input and add_output, (with the same serial id to identify which inputs the sigs are for). this way all parties can send the sigs they can create to their peers (in separate messages or multiple in one, but don't require to send all in one like tx_sig) and relay sigs received from other peers

Meh, this complicates the state machine for no good reason...better to just drop the ordering requirement entirely in that case and just send signatures as soon as you can, don't you think?

Splitting the sigs on a per-input basis wouldn't fundamentally fix the issue. If A is supposed to start in her negotiation with B, if she sends signatures for her inputs but not the inputs she imported from C, why would B start sending their sigs? If there's an ordering requirement, B would wait to receive **all** sigs before sending any of theirs....

I think we don't have a choice, having a strict ordering creates potential deadlocks in multi-party scenarios like the one you describe, so if we want to support that (I guess we do), we must remove that ordering requirement. @niftynei @ddustin WDYT?

Reviewer 6: > better to just drop the ordering requirement entirely in that case and just send signatures as soon as you can

this does not fix it completely, this would work in the case where there are no loops in nodes having a interactive tx session. if nodes have to send all signatures at once at there happens to be a loop they still wait on each other.
when nodes A, B and C batch open a few channels at the same time (and use interactive tx with others to minimize linkability on chain and save on fees) without them knowing of each other, they all wait on the other parties.

i think both are needed:
* no strict ordering
* ability to send signatures separately

Reviewer 2: Can you detail why we'd want to allow loops?
Can that even happen?
It feels to me that it cannot happen because you would end up with a transaction spending the same input twice?
Please detail an example flow to make sure we're on the same page.

Reviewer 2: Can you detail why we'd want to allow loops?
I'm unsure whether they're desirable, and I'm unsure whether they create a deadlock or not.
Please detail an example flow to make sure we're on the same page.

Reviewer 7: > > why not make the signature exchange work like the add_input and add_output, (with the same serial id to identify which inputs the sigs are for). this way all parties can send the sigs they can create to their peers (in separate messages or multiple in one, but don't require to send all in one like tx_sig) and relay sigs received from other peers
> 
> Meh, this complicates the state machine for no good reason...better to just drop the ordering requirement entirely in that case and just send signatures as soon as you can, don't you think?

The spec as it is has you send all signatures you're responsible at once in a variable size tlv https://github.com/lightning/bolts/pull/851/files#diff-ed04ca2c673fd6aabde69389511fa9ee60cb44d6b2ef6c88b549ffaa753d6afeR360-R370 (line 360-370 in 02-peer-protocol.md).

I like this approach and have found it to work well for me. This avoids extra rounds of `tx_signature` back-and-forth that add protocol complications. I can't see any benefit of switching it out for rounds 🤷.

> I think we don't have a choice, having a strict ordering creates potential deadlocks in multi-party scenarios like the one you describe, so if we want to support that (I guess we do), we must remove that ordering requirement. @niftynei @ddustin WDYT?

The conclusion @niftynei and I came to is essentially this:

`tx_signature` griefing is most useful to the side that is sending less sats. By requiring the side with the least sats being added with `tx_add_input` to sign first we eliminate this griefing potential.

```
        +-------+                              +-------+                               +-------+
        |       |-- add_input 200sat --------->|       |<-- add_input 100sat ----------|       |
        |       |<- add_input 10sat -----------|       |--- add_input 10sat ---------->|       |
        |       |<- relay_C_add_input 100sat --|       |-- relay_A_add_input 200sat -->|       |
        |       |                              |       |                               |       |
        |   A   |                              |   B   |                               |   C   |
        |       |                              |       |<-- tx_signature --------------|       |
        |       |<-- tx_signature -------------|       |                               |       |
        |       |--- tx_signature ------------>|       |                               |       |
        |       |                              |       |--- tx_signature ------------->|       |
        +-------+                              +-------+                               +-------+
```

An important note here: when calculating how many sats are on your side you include any inputs you're relaying for others. From the perspective of your peer there is no difference between your own inputs and relayed inputs.

In this scenario C signs first because it sees itself adding 100 sats vs 210 sats coming in.

After receiving C's signatures, B signs next because 110 sats is less than 200 sats coming from A. It forwards this signature plus C's signatures to A in a single `tx_signature` message.

Then A signs and sends `tx_signature` along to B.

Finally, B bundles A's signatures with its own and sends them to C.

> having a strict ordering creates potential deadlocks in multi-party scenarios like the one you describe, so if we want to support that (I guess we do), we must remove that ordering requirement.

Not having an ordering requirement is what actually introduces the deadlock potential -- the incentive is to always sign last. Having a strict ordering of who must sign first that is whomever has the least to lose is designed to mitigate it.

Author: > i imagine 3 nodes (A, B and C) setting up a triangle of channels where every node contributes the same amount of sats.

fwiw this is impossible, as it'd break the rule that you fail the interactive tx negotiation if a node sends you an input that's yours. (as C would ostensibly do to A).

You could make 2 channels (A-B-C) but not a circle (A-B-C-A).

Reviewer 2: I thought about this some more, and I agree with @niftynei and @ddustin. @antonilol feel free to share an example scenario in case I misunderstood your point, but I don't think what you wanted to achieve would work, regardless of signatures ordering. Loops are simply not possible, as only one node at a time can batch multiple channel opens into a single tx, otherwise there would be a deadlock waiting for `tx_complete` (so even before we reach the signatures step). 

Here is why:

- the full transaction must be the same in all parallel sessions (otherwise signatures can't be transferred from one session to another)
- so a node that wants to batch multiple sessions in a single tx needs to wait for their peers to send `tx_complete` before sending their own, because they need to have the full set of inputs/outputs of each peer to transfer them to the other sessions
- this batching node will probably "cheat" by adding and removing a dummy input until they receive `tx_complete` from their peers
- if both nodes in a session do that, it will never converge and nobody will send `tx_complete`
- when that happens, negotiation will fail after 4096 iterations of adding an input or output

Also, the ordering requirement based on amount works well for this kind of batching. Let's consider B batching a session with A and a session with C. From A's point of view, B is adding `amount_BA + amount_CB` to the channel. From C's point of view, B is adding `amount_BC + amount_AB` to the channel. For a deadlock at the signing step to occur, we would need to have both `amount_BA + amount_CB <= amount_AB` and `amount_BC + amount_AB <= amount_CB`, which implies `amount_BA + amount_CB <= amount_CB - amount_BC` => `amount_BA <= -amount_BC` which is only possible if B contributes nothing to the channels...so as long as the coordinator contributes a non-zero amount to at least one channel, there is a channel where they won't start sending their sigs, so they won't be blocked. Let me know if you want me to formalize it better (I believe this generalizes to N sessions but haven't written it down yet, will do today).

Maybe the spec should mention that only the initiator should try to batch multiple sessions, otherwise there's a risk of deadlock?

Reviewer 2: > I believe this generalizes to N sessions but haven't written it down yet, will do today

It does generalize to N sessions, as long as the coordinator contributes a non-zero amount to each channel, there can't be a deadlock. I can share the (small) proof if you're interested.

Author: fwiw I believe @ddustin and I independently reached this same conclusion wrt "no deadlock as long as the coordinator contributes funds."

Reviewer 6: > @antonilol feel free to share an example scenario in case I misunderstood your point, but I don't think what you wanted to achieve would work, regardless of signatures ordering. Loops are simply not possible, as only one node at a time can batch multiple channel opens into a single tx, otherwise there would be a deadlock waiting for `tx_complete` (so even before we reach the signatures step).

when making it i came to the same `tx_complete` deadlock. i would work though if all collaborating parties know that they all work on the same tx and have some other medium to send messages (to coordinate who sets it up and to not make loops)

how should accidental loops be handled? multiple nodes that work on the same tx (but not all) will notice it when their in/output gets sent back and fail the negotiation. others will notice just that their peer disconnected after they sent a in/output (that was theirs).

Reviewer 2: > how should accidental loops be handled?

If only the initiator does that kind of batching, there is no risk of accidental loops. I would simply recommend against trying to batch when you're not the initiator, you'll always run the risk of creating those loops otherwise.

> how should accidental loops be handled?

When a node detects a loop, they will send `tx_abort` (with a custom message that may explain why they're aborting) to cancel everything.

If there is a deadlock before `tx_complete`, the protocol will abort after 4096 iterations as well, and this is a good indication that both sides of an interactive-tx session are trying to batch at the same time, which cannot work.

> sessions to collectively construct a single transaction. This preserves

Reviewer 7: Looks right except (perhaps a typo) that A & B have to take turns sending messages so A sends tx_complete first.

```
        +-------+                          +-------+                          +-------+
        |       |--(1)- tx_add_input_a --->|       |--(1)- tx_add_input_b --->|       |
        |       |<-(2)- tx_add_input_b ----|       |<-(2)- tx_complete -------|       |
        |       |--(3)- tx_add_output_a -->|       |--(3)- tx_add_input_a  -->|       |
        |       |<-(4)- tx_add_output_b ---|       |<-(4)- tx_complete -------|       |
        |   A   |--(5)- tx_complete ------>|   B   |--(5)- tx_add_output_a -->|   C   |
        |       |<-(6)- tx_complete -------|       |<-(6)- tx_complete -------|       |
        |       |                          |       |--(7)- tx_add_output_b -->|       |
        |       |                          |       |<-(8)- tx_complete -------|       |
        |       |                          |       |--(9)- tx_complete ------>|       |
        +-------+                          +-------+                          +-------+
```

> sessions to collectively construct a single transaction. This preserves

Reviewer 8: Nice catch, thanks!

  - MUST use a unique `serial_id` for each input currently added to the
    transaction
  - MUST NOT re-transmit inputs it has received from the peer
  - if is the *initiator*:
    - MUST send even `serial_id`s
  - if is the *non-initiator*:
    - MUST send odd `serial_id`s

The receiving node:
  - MUST add all received inputs to the transaction
  - MUST fail the negotiation if:
    - the `prevtx` and `prevtx_vout` are identical to a previously added
      (and not removed) input's
    - `prevtx` is not a valid transaction
    - `prevtx_vout` is greater or equal to the number of outputs on `prevtx`
    - the `scriptPubKey` of the `prevtx_vout` output of `prevtx` is not exactly a 1-byte push opcode (for the numeric values `0` to `16`) followed by a data push between 2 and 40 bytes
    - the `serial_id` is already included in the transaction
    - the `serial_id` has the wrong parity
    - if has received 4096 `tx_add_input` messages during this negotiation

#### Rationale

Each node must know the set of the transaction inputs. The *non-initiator*
MAY omit this message.

`serial_id` is a randomly chosen number which uniquely identifies this input.

> Inputs in the constructed transaction are sorted by `serial_id`.

Reviewer 9: Why not use BIP69 ordering?

Using this random ordering may become a marker to identify lightning channels on chain.

Reviewer 2: The reason to use that ordering instead of BIP 69 is to potentially allow usage of `sighash_single` which requires participants to control the order of inputs/outputs.

It's not trivial to effectively control this though: if you use serial IDs 0 and 2 for and input and output that you want to sign using `sighash_single`, there is a risk that your peer will add an output with serial ID 1, which messes that up. You may be able to rectify this using `tx_remove_input` and `tx_remove_output` to insert them at a different place, but that won't work in every case. But in some higher-level protocols that specify how serial IDs must be assigned to allow exotic sighash constructions, this would come in handy and would be impossible with BIP 69. Does that seem like a reasonable trade-off?

Reviewer 9: It's hard for me to judge -- I'm not familiar with all the applications of the `interactive-tx` protocol.

At least for dual funding, I don't see a reason to use `sighash_single`.

Reviewer 6: The only application i can think of is attaching htlc 1st level txs

Reviewer 2: My personal preference is to favor flexibility for protocols that will build on top of `interactive-tx`, because we don't want to restrict what can be built. We can always later add an optional tlv to request using a different ordering (e.g. bip 69) if we feel it's important.

Reviewer 2: > At least for dual funding, I don't see a reason to use sighash_single.

But a dual funding transaction can be batched with another type of transaction (splicing, coinjoin, etc) and you don't even have to know about it (which is great). We don't want to make that impossible in my opinion.

`prevtx` is the serialized transaction that contains the output
this input spends. Used to verify that the input is non-malleable.

`prevtx_vout` is the index of the output being spent.

`sequence` is the sequence number of this input.  Must be less than

> 4294967294 (0xFFFFFFFE). See BIP125.

Reviewer 6: a transaction signals rbf if **at least 1 input** has `nSequence < 0xfffffffe`
only forcing one input's nSequence to be below 0xfffffffe feels more correct

also, a link to bip125 could be added here:
`https://github.com/bitcoin/bips/blob/master/bip-0125.mediawiki#summary`

Reviewer 2: I'm not sure we should mention BIP 125, as bitcoin core is in the process of removing references to it since what's actually implemented today doesn't match BIP 125...

    * [`u64`:`sats`]
    * [`u16`:`scriptlen`]
    * [`scriptlen*byte`:`script`]

#### Requirements

Either node:
  - MAY omit this message

The sending node:
  - MUST add all sent outputs to the transaction
  - if is the *initiator*:
    - MUST send even `serial_id`s
  - if is the *non-initiator*:
    - MUST send odd `serial_id`s

The receiving node:
  - MUST add the specified output to the transaction
  - MUST accept P2WSH, P2WPKH, P2TR `script`s
  - MAY fail the negotiation if `script` is non-standard
  - MUST fail the negotiation if:
    - the `serial_id` is already included in the transaction
    - the `serial_id` has the wrong parity
    - it has received 4096 `tx_add_output` messages during this negotiation
    - the `sats` amount is less than the `dust_limit`
    - the `sats` amount is greater than 2,100,000,000,000,000 (`MAX_MONEY`)

#### Rationale

Each node must know the set of the transaction outputs.

`serial_id` is a randomly chosen number which uniquely identifies this output.
Outputs in the constructed transaction are sorted by `serial_id`.

`sats` is the satoshi value of the output.

`script` is the scriptPubKey for the output (with its length omitted).

> It's left undefined if you accept other standard outputs such as `OP_RETURN`

Reviewer 10: Not sure I quite get what this is saying with "It's left undefined if you...". It kind of sounds a little like it's an optional field from this phrasing, but I'm not sure how that makes sense. And it's not explicit who the "you" is here. Sorry if I'm missing something lol.

The maximum inputs and outputs are capped at 252. This effectively fixes
the byte size of the input and output counts on the transaction to one (1).

### The `tx_signatures` Message

1. type: 71 (`tx_signatures`)
2. data:
    * [`channel_id`:`channel_id`]
    * [`sha256`:`txid`]
    * [`u16`:`num_witnesses`]
    * [`num_witnesses*witness_stack`:`witness_stack`]

1. subtype: `witness_stack`
2. data:
    * [`u16`:`num_input_witness`]
    * [`num_input_witness*witness_element`:`witness_element`]

1. subtype: `witness_element`
2. data:
    * [`u16`:`len`]
    * [`len*byte`:`witness`]

#### Requirements

The sending node:
  - if it has the lowest total satoshis contributed, as defined
    by total `tx_add_input` values:
    - MUST transmit their `tx_signatures` before the peer
  - MUST order the `witness_stack`s by the `serial_id` of the input they
    correspond to
  - number of `witness_stack`s MUST equal the number of inputs they added

The receiving node:
  - MUST fail the negotiation if:
    - if they have not sent a `tx_signatures` message and they are the
      peer with the lowest total satoshis contributed, as defined by

>       `tx_add_input` values

Reviewer 2: Right, I fully agree, there is no reason to fail if we receive our counterparty's signatures early, this requirement should be removed. Eclair already supports receiving early `tx_signatures` (since as you mention, it's clearly a good thing for our node to receive the remote sigs).

    - the message contains an empty `witness_stack`
    - the number of `witness_stack`s does not equal the number of inputs
      added by the sending node
    - the `txid` does not match the txid of the transaction
  - SHOULD apply the `witness`es to the transaction and broadcast it
  - MUST reply with their `tx_signatures` if not already transmitted

#### Rationale

The peer with the lowest total of inputs amounts must transmit its

> `tx_signatures` first. If both peers have contributed equal input amounts,

Reviewer 9: There is a griefing attack that results from this `tx_signatures` ordering.

Suppose you have configured your node to match funds from all `open_channel2`s up to a certain amount.

A malicious node can repeatedly initiate negotiations with your node, using UTXOs they don't own to construct the funding tx. The malicious node can ensure they have the larger contribution to the funding tx, forcing your node to send `tx_signatures` first. The malicious node then refuses to send their `tx_signatures`, forcing lockup of your funds until your node times out and double-spends them.

In this way, malicious node(s) can force your node to either lockup all its available funds or  repeatedly waste tx fees from double-spends, eventually consuming all your available funds.

Would it make more sense to require the opener to send `tx_signatures` first?  This would allow the acceptor to grief the opener, but the opener can easily go elsewhere if they get burned.

Reviewer 2: > Suppose you have configured your node to match funds from all open_channel2s up to a certain amount.

That is a strong assumption, a node using that behavior knows that they are taking some risk and must implement some kind of DoS protection anyway. I wouldn't recommend such a strategy anyway.

> In this way, malicious node(s) can force your node to either lockup all its available funds or repeatedly waste tx fees from double-spends, eventually consuming all your available funds.

Not really, because the attacked node would just reuse the same utxos in another channel opening flow, which provides the double-spending part "for free" (no fees paid). The timeout for receiving `tx_signatures` can be very low, as there's no reason for an honest node to withhold their `tx_signatures` for a long time.

> Would it make more sense to require the opener to send tx_signatures first?

No, I think this would be worse in terms of protocol flexibility as this would remove the ability for an opener to batch funding transactions by coordinating multiple sessions in parallel. It's trivial to avoid the griefing attack you mention by just contributing 1 sat less than the opener (if you insist on matching their contribution), so I don't think we should be very concerned about this.

Reviewer 9: > That is a strong assumption, a node using that behavior knows that they are taking some risk and must implement some kind of DoS protection anyway. I wouldn't recommend such a strategy anyway.

The exact policy isn't too important, since the attacker can defeat any of them.  For example, here's CLN's available policies:

- match -- Contribute policy_mod percent of their requested funds. Valid policy_mod values are 0 to 200. If this is a channel lease request, we match based on their requested funds. If it is not a channel lease request (and lease_only is false), then we match their funding amount. Note: any lease match less than 100 will likely fail, as clients will not accept a lease less than their request.
- available -- Contribute policy_mod percent of our available node wallet funds. Valid policy_mod values are 0 to 100.
- fixed -- Contributes a fixed policy_mod sats to v2 channel open requests.

For any of these, the attacker can keep doing `tx_add_input` to maintain a larger contribution.

But I am curious... How do you see dual funding being used without some fixed policy like these?  

> Not really, because the attacked node would just reuse the same utxos in another channel opening flow, which provides the double-spending part "for free" (no fees paid). The timeout for receiving `tx_signatures` can be very low, as there's no reason for an honest node to withhold their `tx_signatures` for a long time.

Right, we can productively double-spend, which helps against the attack.  But we still need to remember every pending channel using a UTXO until one of them confirms or we double spend.  If the attacker is blasting us with pending channels that will *never* confirm, it will eat up our memory/disk space until we crash or are forced to double-spend.

> No, I think this would be worse in terms of protocol flexibility as this would remove the ability for an opener to batch funding transactions by coordinating multiple sessions in parallel. It's trivial to avoid the griefing attack you mention by just contributing 1 sat less than the opener (if you insist on matching their contribution), so I don't think we should be very concerned about this.

I think you meant 1 sat *more* than the opener.

It's not so trivial.  The `tx_signatures` order is determined by value of inputs contributed, not by the `funding_sats` specified in `open_channel2`/`accept_channel2`.  So the attacker can keep doing `tx_add_input` to maintain a larger contribution (remember they have the entire UTXO set to pick from).

Reviewer 9: > But that's something you have to do all the time, right? You always need to watch unconfirmed channels until they either confirm or are double-spent (or, if you didn't participate in funding, you can just forget them after a while). You will likely be running out of utxos to use in funding before memory becomes a problem, so I'm not sure you should be so concerned about this.

In the v1 flow, the acceptor can safely forget the pending channel at any point, since they didn't contribute any funds.  In v2, a dual-funding acceptor can't forget until an input in the pending channel is double spent.  Since the acceptor is attempting to productively double-spend their UTXOs, they may not actually run out -- they keep reusing the same UTXOs for each dual-funding request, consuming additional memory each time for the new pending channel.

> I'm not sure what is best, we could also make the opener sign first by default, with an optional TLV in `open_channel2` that would let them say "Hey, I'm trying to batch, please be a gent and send tx_signatures first this time"? And the non-initiator would contribute to only a handful of those per-day (or any restricting policy) to restrict their exposure. If this is used in the context of liquidity ads, since the non-initiator gets a fee in the end, it's probably a good enough incentive to take some risk and sign first? What do you think? Curious to know what @niftynei thinks about this as well.

It seems for simple dual funding, opener signing first is the simplest.  But if we're going to optimize for the 2-party case, the whole interactive-tx session is unnecessary anyway (we could just send all inputs/outputs once).

I cringe thinking about the complexity, but maybe something like PoDLEs are the ultimate solution here...

Reviewer 3: From my understanding this issue of lack of atomicity of signatures exchanges, introducing DoS vectors or economic disequilibrium of the collaborative transaction has been already hit within the Discreet Log Contract specification effort. Analogously, two participants are contributing collateral UTXOs to the same off-chain contract state over multiple Lightning hops. One idea brought out to solve the issue was "Barrier Escrow", a third-party guaranteeing the release of participants signature scalars: https://suredbits.com/payment-points-and-barrier-escrows/. The trust model itself stays to be studied.
 
About PoDLEs, I think they could be a significant v2 channels coins griefing vector rate-limiting, in the sense you can share the non-honored UTXO commitments among LSP (let's say over gossips). I don't think this obliterate the risks for an individual LSP in face of an adversary able to renew its stock of griefing UTXOs in period of low-fees. 

Note, I don't think this is something we should aim for either v0.1 or v0.2 of the interactive transaction construction protocol, however that might be long-term set of solutions.

the peer with the first `node_id` (sorted lexicographically) sends their
`tx_signatures` first.  This gives in a strict ordering of transmission
for any multiparty tx collaboration.

`witness` is the data for a witness element in a witness stack.

> Witness elements should *not* include their length.

Reviewer 9: Or, even simpler:  we can rename `witness` to `data` and refer to it as "`witness_element.data` should *not* include a length encoding".

required fee, e.g. a multisig witness stack whose weight exceeds 107.

### The `tx_init_rbf` Message

This message initiates a replacement of the transaction after it's been
completed.

1. type: 72 (`tx_init_rbf`)
2. data:
   * [`channel_id`:`channel_id`]
   * [`u32`:`locktime`]
   * [`u32`:`feerate`]
   * [`tx_init_rbf_tlvs`:`tlvs`]

1. `tlv_stream`: `tx_init_rbf_tlvs`
2. types:
    1. type: 0 (`funding_output_contribution`)
    2. data:
        * [`tu64`:`satoshis`]

#### Requirements

The sender:
  - MUST set `feerate` greater than or equal to 25/24 times the `feerate`
    of the previously constructed transaction, rounded down.

The recipient:
  - MUST respond either by failing the negotiation or with `tx_ack_rbf`.
  - MUST fail the negotiation if:
    - the `feerate` is not greater than 25/24 times `feerate` of the last
      successfully constructed transaction
  - MAY fail the negotiation for any reason

#### Rationale

`feerate` is the feerate this transaction will pay. It must be at least
1/24 greater than the last used `feerate`, rounded down to the nearest
satoshi to ensure there is progress.

E.g. if the last `feerate` was 520, the next sent `feerate` must be 541

> (520 * 25 / 24 = 541.667, rounded down to 541).

Reviewer 6: this allows for creation of txs that cant be broadcast when the initial fee rate is below `24*minrelayfee` sat/vB

i think it also shouldn't be allowed to initiate rbf with a feerate that does not cover the retransmission cost
the retransmission cost is `minrelayfee` * `virtual size` rounded up in sats

minrelayfee is 1 sat/vB at the moment but can change in the future (`bitcoin-cli getmempoolinfo`, returns in BTC/kvB)

source:
https://github.com/bitcoin/bips/blob/master/bip-0125.mediawiki#implementation-details number 4

Reviewer 2: I don't think the lightning spec should take those details into account. If bitcoin changes relay policies, we don't want to have to change the spec, only implementations should change. IMO it's ok to be somewhat hand-wavy regarding RBF, stating something general like "implementations are free to reject any RBF attempt that may fail to propagate".

I think it's good that the spec requires a minimal % increase, as it's protects against DoS, but this is just a lower bound.

Also, it's not such a big deal if we generate a tx that doesn't propagate because it doesn't meet the bitcoin network's relay policies, we'll just RBF again after some time and eventually we'll have something that relays and confirms.

Reviewer 6: > I don't think the lightning spec should take those details into account. If bitcoin changes relay policies, we don't want to have to change the spec

yeah i get that, `minrelayfee` could be put in the spec (as text, not 1 sat/vB), and even implementation dont have to change if the backend supplies this information. but this can get overly complicated because some implementation may not use a backend that supplies this info, and 2 nodes could have a different value for this

> IMO it's ok to be somewhat hand-wavy regarding RBF, stating something general like "implementations are free to reject any RBF attempt that may fail to propagate".
> 
> I think it's good that the spec requires a minimal % increase, as it's protects against DoS, but this is just a lower bound.
> 
> Also, it's not such a big deal if we generate a tx that doesn't propagate because it doesn't meet the bitcoin network's relay policies, we'll just RBF again after some time and eventually we'll have something that relays and confirms.

true

Reviewer 7: > I think it's good that the spec requires a minimal % increase, as it's protects against DoS, but this is just a lower bound.
> 
> 
> 
> Also, it's not such a big deal if we generate a tx that doesn't propagate because it doesn't meet the bitcoin network's relay policies, we'll just RBF again after some time and eventually we'll have something that relays and confirms.

I agree with this. It's worth keeping in mind in any case that splicing (which is using the same protocol) may never have anything that confirms and has to maintain HTLC sig duplicates for every RBF. So the cost of RBF spam there is higher.

Reviewer 7: The goal of splicing is lowest possible fees though so we want low fee attempts. Given that & it's negotiated between two peers who both bear the burden -- I suspect it's totally worth the extra overhead of low RBF bump minimums.

If the previous transaction confirms in the middle of an RBF attempt,
the attempt MUST be abandoned.

`funding_output_contribution` is the amount of satoshis that this peer
will contribute to the funding output of the transaction, when there is
such an output. Note that it may be different from the contribution

> made in the previously completed transaction.

Reviewer 6: nit: duplicate of lines +482 to +485

### The `tx_ack_rbf` Message

1. type: 73 (`tx_ack_rbf`)
2. data:
   * [`channel_id`:`channel_id`]
   * [`tx_ack_rbf_tlvs`:`tlvs`]

1. `tlv_stream`: `tx_ack_rbf_tlvs`
2. types:
    1. type: 0 (`funding_output_contribution`)
    2. data:
        * [`tu64`:`satoshis`]

#### Requirements

The recipient:
  - MUST either fail the negotiation or transmit a `tx_add_input` message,
    restarting the interactive tx collaboration protocol.

#### Rationale

`funding_output_contribution` is the amount of satoshis that this peer

> will contribute to the funding output of the transaction, when there is

Reviewer 6: nit: "is willing to" sounds better to me
```suggestion
is willing to contribute to the funding output of the transaction, when there is
```

Author: `is willing to` has a different meaning than `will`.

such an output. Note that it may be different from the contribution
made in the previously completed transaction.

It's recommended that a peer, rather than fail the RBF negotiation due to
a large feerate change, instead sets their `funding_output_contribution` to

> zero, and decline to participate further in the transaction.

Reviewer 9: Why?

### The `tx_abort` Message

1. type: 74 (`tx_abort`)
2. data:
   * [`channel_id`:`channel_id`]
   * [`u16`:`len`]
   * [`len*byte`:`data`]

The `len` field indicates the number of bytes in the immediately following field.

#### Requirements

A sending node:
  - MUST NOT have already transmitted `tx_signatures`
  - SHOULD forget the current negotiation and reset their state.

A receiving node:
  - if they have already sent `tx_signatures` to the peer:
    - MUST NOT forget the channel until any inputs to the negotiated tx
      have been spent.
  - if they have not sent `tx_signatures`:
    - SHOULD forget the current negotiation and reset their state.

#### Rationale

A receiving node, if they've already sent their `tx_signatures` has no guarantee
that the transaction won't be signed and published by their peer. They must remember
the transaction and channel (if appropriate) until the transaction is no longer
eligible to be spent (i.e. any input has been spent in a different transaction).

The `tx_abort` message allows for the cancellation of an in progress negotiation,

> and a return to the initial starting state.

Reviewer 2: Because you don't want to fail the channel when you are in an `interactive-tx` session to RBF the funding tx for example. In that case you just want to fail the RBF `interactive-tx` session, but the previous funding txs may confirm so you don't want to fail the channel.

> and a return to the initial starting state.

Reviewer 9: Makes sense.  We should explain this under Rationale, and probably define the term "fail the negotiation" (as opposed to "fail the channel") in the Glossary.

Do we always send `tx_abort` instead of `error` in v2?  Only during the `interactive-tx` part?  Only during RBF?

Do we send `tx_abort` for some errors, and `error` for others?

> and a return to the initial starting state.

Reviewer 2: > Do we always send tx_abort instead of error in v2? Only during the interactive-tx part? Only during RBF?
> Do we send tx_abort for some errors, and error for others?

You should send `tx_abort` when you want to abort an `interactive-tx` protocol, but don't necessarily want to reject the channel.
You should send `error` when you want to reject the whole channel opening flow.

Those are really two different intents, and you'll see in the implementation that it's somewhat obvious which one you should use depending on what internal error is hit.

    - the witnesses are non-standard
  - SHOULD apply the `witnesses` to the transaction and broadcast it
  - MUST reply with their `tx_signatures` if not already transmitted

#### Rationale

A strict ordering is used to decide which peer sends `tx_signatures` first.
This prevents deadlocks where each peer is waiting for the other peer to
send `tx_signatures`, and enables multiparty tx collaboration.

`witness_data` is the data for a witness element in a witness stack, not
prefixed with its length (since it is already specified in the `len` field).

While the `minimum fee` is calculated and verified at `tx_complete` conclusion,
it is possible for the fee for the exchanged witness data to be underpaid.
It is the responsibility of the sending peer to correctly account for the
required fee, e.g. a multisig witness stack whose weight exceeds 107.

### The `tx_init_rbf` Message

This message initiates a replacement of the transaction after it's been
completed.

1. type: 72 (`tx_init_rbf`)
2. data:
   * [`channel_id`:`channel_id`]
   * [`u32`:`locktime`]
   * [`u32`:`feerate`]
   * [`tx_init_rbf_tlvs`:`tlvs`]

1. `tlv_stream`: `tx_init_rbf_tlvs`
2. types:

>     1. type: 0 (`funding_output_contribution`)

Reviewer 9: If `funding_output_contribution` is missing, do we assume the contribution is the same as before?

Reviewer 2: If `funding_output_contribution` is missing, it means the peer is not contributing to the channel (it's equivalent to having set a value of `0`). That should be explicitly mentioned.

Reviewer 2: Fixed in https://github.com/niftynei/lightning-rfc/pull/10

Author: merged!

    2. data:
        * [`tu64`:`satoshis`]

#### Requirements

The sender:
  - MUST set `feerate` greater than or equal to 25/24 times the `feerate`
    of the previously constructed transaction, rounded down.

The recipient:
  - MUST respond either with `tx_abort` or with `tx_ack_rbf`
  - MUST respond with `tx_abort` if:

>     - the `feerate` is not greater than 25/24 times `feerate` of the last

Reviewer 9: ```suggestion
    - the `feerate` is not greater than or equal to 25/24 times `feerate` of the last
```

Reviewer 2: Fixed in https://github.com/niftynei/lightning-rfc/pull/10

Author: merged!

    - there are more than 252 inputs
    - there are more than 252 outputs
    - the estimated weight of the tx is greater than 400,000 (`MAX_STANDARD_TX_WEIGHT`)

#### Rationale

To signal the conclusion of exchange of transaction inputs and outputs.

Upon successful exchange of `tx_complete` messages, both nodes
should construct the transaction and proceed to the next portion of the
protocol. For channel establishment v2, exchanging commitment transactions.

For the `minimum fee` calculation see [BOLT #3](03-transactions.md#calculating-fees-for-collaborative-transaction-construction).

The maximum inputs and outputs are capped at 252. This effectively fixes
the byte size of the input and output counts on the transaction to one (1).

### The `tx_signatures` Message

1. type: 71 (`tx_signatures`)
2. data:
    * [`channel_id`:`channel_id`]
    * [`sha256`:`txid`]
    * [`u16`:`num_witnesses`]
    * [`num_witnesses*witness_stack`:`witnesses`]

1. subtype: `witness_stack`
2. data:
    * [`u16`:`num_witness_elements`]
    * [`num_witness_elements*witness_element`:`witness_elements`]

1. subtype: `witness_element`
2. data:
    * [`u16`:`len`]

>     * [`len*byte`:`witness_data`]

Reviewer 10: Sorry for asking late again, but missed this. Is there a reason we don't just say the witnesses are each serialized as in bitcoin consensus and not have these subtypes? It makes encoding/decoding a little easier.

Reviewer 10: Or rather, as [proposed](https://github.com/lightningdevkit/rust-lightning/pull/1794#discussion_r1095155072) by Wilmer, keep the subtypes, but define them as:

```
1. subtype: `witness_stack`
2. data:
    * [`compactsize`:`num_witness_elements`]
    * [`num_witness_elements*witness_element`:`witness_elements`]
1. subtype: `witness_element`
2. data:
    * [`compactsize`:`len`]
    * [`len*byte`:`witness_data`]
```

which means we'll need to add `compactsize` as a fundamental type.

It's still explicit which is good, and we can use existing consensus (de)serialisation.

Reviewer 2: Meh, this would create inconsistencies with the rest of the lightning data types. I'd favor consistency inside lightning and ignoring the bitcoin consensus serialization (on the eclair side, we're not exposed to the consensus serialization here, and I believe it's the same for cln, so I don't see a compelling reason to break consistency for this).

Reviewer 10: Cool, not a major pain point. :)

1/24 greater than the last used `feerate`, rounded down to the nearest
satoshi to ensure there is progress.

E.g. if the last `feerate` was 520, the next sent `feerate` must be 541
(520 * 25 / 24 = 541.667, rounded down to 541).

If the previous transaction confirms in the middle of an RBF attempt,
the attempt MUST be abandoned.

`funding_output_contribution` is the amount of satoshis that this peer
will contribute to the funding output of the transaction, when there is
such an output. Note that it may be different from the contribution
made in the previously completed transaction. If omitted, the sender is
not contributing to the funding output.

### The `tx_ack_rbf` Message

1. type: 73 (`tx_ack_rbf`)
2. data:
   * [`channel_id`:`channel_id`]
   * [`tx_ack_rbf_tlvs`:`tlvs`]

1. `tlv_stream`: `tx_ack_rbf_tlvs`
2. types:
    1. type: 0 (`funding_output_contribution`)
    2. data:
        * [`tu64`:`satoshis`]

#### Requirements

The sender:
  - If it contributes to the transaction's funding output:
    - MUST set `funding_output_contribution`

The recipient:
  - MUST respond with `tx_abort` or with a `tx_add_input` message,

>     restarting the interactive tx collaboration protocol.

Reviewer 2: I think it makes the logic more complex if RBF attempts start with the previous transaction and add/remove inputs, it's simpler and more general to instead start from scratch and after exchanging `tx_complete`, simply verify that the new tx correctly double-spends previous attempts.

Eclair is currently not using `tx_remove_*` messages, but I think they can be useful for protocols that create transactions by batching multiple interactive-tx sessions: if you're trying to batch channel opens with Alice, Bob and Carol, but Bob drops off during the interactive-tx session, you'll want to remove the inputs/outputs you added for him from the txs you're creating with Alice and Carol.

>     restarting the interactive tx collaboration protocol.

Reviewer 10: Thanks, this really clears it up for me!

pubkey corresponding to the funding output nothing prevents duplicative channel
ids.

### `channel_id`, v2

For channels established using the v2 protocol, the `channel_id` is the
`SHA256(lesser-revocation-basepoint || greater-revocation-basepoint)`,
where the lesser and greater is based off the order of the basepoint.

When sending `open_channel2`, the peer's revocation basepoint is unknown.
A `temporary_channel_id` must be computed by using a zeroed out basepoint
for the non-initiator.

When sending `accept_channel2`, the `temporary_channel_id` from `open_channel2`
must be used, to allow the initiator to match the response to its request.

#### Rationale

The revocation basepoints must be remembered by both peers for correct
operation anyway. They're known after the first exchange of messages,
obviating the need for a `temporary_channel_id` in subsequent messages.
By mixing information from both sides, they avoid `channel_id` collisions,

> and they remove the dependency on the funding txid.

Reviewer 3: An on-chain transaction ID binds to the following serialization format:

`[nVersion][txins][txouts][nLockTime]`

For legacy channels (v1 funding flow), the `txins`, the `txouts` (apart of the funding output) and the `nLockTime` (for the value in the past) constitute an information space that can be freely tweaked by the funding counterparty. This information space tweaking can be iterated until a collision for a known v2 channel id is found (same process than finding proof-of-work for miner). Once a collision is found this id could be announced through the v1 funding flow and provoke “undefined behavior” in the counterparty’s LN implementation (e.g force-close of the colliding channels).

Of course, this concern is not practical as SHA256 has preimage resistance as a hash function. Still, if more channels ids are introduced in the future, implementations could recommend to prefix their channel id contexts. 
This could be of interest today, as BOLT3 does not have recommendation on `revocation_basepoint` derivation, and multi-party curve point construction (e.g MuSig2) opens the door to all the attack vectors affecting distributed key generation.

        |       |--(9)- tx_complete ----->|       |
        |       |<-(10) tx_complete ------|       |
        +-------+                         +-------+

### The `tx_add_input` Message

This message contains a transaction input.

1. type: 66 (`tx_add_input`)
2. data:
    * [`channel_id`:`channel_id`]
    * [`u64`:`serial_id`]
    * [`u16`:`prevtx_len`]
    * [`prevtx_len*byte`:`prevtx`]
    * [`u32`:`prevtx_vout`]
    * [`u32`:`sequence`]

#### Requirements

The sending node:
  - MUST add all sent inputs to the transaction
  - MUST use a unique `serial_id` for each input currently added to the
    transaction
  - MUST set `sequence` to be less than or equal to 4294967293 (`0xFFFFFFFD`)
  - MUST NOT re-transmit inputs it has received from the peer
  - if is the *initiator*:
    - MUST send even `serial_id`s
  - if is the *non-initiator*:
    - MUST send odd `serial_id`s

The receiving node:
  - MUST add all received inputs to the transaction
  - MUST fail the negotiation if:
    - `sequence` is set to `0xFFFFFFFE` or `0xFFFFFFFF`
    - the `prevtx` and `prevtx_vout` are identical to a previously added
      (and not removed) input's

>     - `prevtx` is not a valid transaction

Reviewer 2: I'm not sure you would want to necessarily reject those? Even if `prevtx` has an `nLockTime` that is a few blocks away, it could still be ok, it just means you'll need to wait a bit before you publish the funding transaction, but that could be acceptable, right? I think each implementation will choose its own policy here, as there are many distinct scenarios.

>     - `prevtx` is not a valid transaction

Reviewer 3: Correct implementation-level, one could negotiate collaborative transaction and use timelocks to “block” its counterparty from broadcasting transactions until X, as understood by a yet-to-be-defined liquidity negotiation protocol.

    - `prevtx_vout` is greater or equal to the number of outputs on `prevtx`
    - the `scriptPubKey` of the `prevtx_vout` output of `prevtx` is not exactly a 1-byte push opcode (for the numeric values `0` to `16`) followed by a data push between 2 and 40 bytes
    - the `serial_id` is already included in the transaction
    - the `serial_id` has the wrong parity

>     - if has received 4096 `tx_add_input` messages during this negotiation

Reviewer 2: Do you really expect that to happen? The way I see it, an `interactive-tx` session should be fast. Not necessarily in the dual funding case, but at least for splicing, we don't want the channel to be inactive for too long because of a slow `interactive-tx` session (because while the channel is inactive, you're not earning routing fees). In `eclair` we have an explicit 60 seconds timeout: if the `interactive-tx` construction isn't finished before that, we disconnect to discard it entirely.

I don't think it makes sense then to consider fee variations in such a small time interval, that would IMO add unnecessary complexity.

>     - if has received 4096 `tx_add_input` messages during this negotiation

Reviewer 3: > Do you really expect that to happen?

Yes. In a world where you have a competition between on-chain fees (based blockspace demand) and off-chain fees (based on channel topology) you will always have a price equilibrium seeked by economically rational actors.

Though I don’t see it happening in the coming years, maybe not even in the coming decade, so all good.

#### Rationale

Each node must know the set of the transaction inputs. The *non-initiator*
MAY omit this message.

> `serial_id` is a randomly chosen number which uniquely identifies this input.

Reviewer 2: I think you can do whatever you want here, an implementation can choose to derive that from the spent outpoint (and add 1 for oddness if necessary) if it is simpler.

> `serial_id` is a randomly chosen number which uniquely identifies this input.

Reviewer 3: Yes correct this can be implementation-level.

> Inputs in the constructed transaction are sorted by `serial_id`.

Reviewer 10: nit: Turns out this was easy to miss in the rendered markdown as the line just follows on from the previous line. Can we make it "... transaction MUST be sorted by..." so it's clearer to parse?

### The `tx_add_output` Message

This message adds a transaction output.

1. type: 67 (`tx_add_output`)
2. data:
    * [`channel_id`:`channel_id`]
    * [`u64`:`serial_id`]
    * [`u64`:`sats`]
    * [`u16`:`scriptlen`]
    * [`scriptlen*byte`:`script`]

#### Requirements

Either node:
  - MAY omit this message

The sending node:
  - MUST add all sent outputs to the transaction
  - if is the *initiator*:
    - MUST send even `serial_id`s
  - if is the *non-initiator*:
    - MUST send odd `serial_id`s

The receiving node:
  - MUST add all received outputs to the transaction
  - MUST accept P2WSH, P2WPKH, P2TR `script`s
  - MAY fail the negotiation if `script` is non-standard
  - MUST fail the negotiation if:
    - the `serial_id` is already included in the transaction
    - the `serial_id` has the wrong parity
    - it has received 4096 `tx_add_output` messages during this negotiation
    - the `sats` amount is less than the `dust_limit`
    - the `sats` amount is greater than 2,100,000,000,000,000 (`MAX_MONEY`)

#### Rationale

Each node must know the set of the transaction outputs.

`serial_id` is a randomly chosen number which uniquely identifies this output.

> Outputs in the constructed transaction are sorted by `serial_id`.

Reviewer 10: Same here for nit: "... transaction MUST be sorted by...".

`sats` is the satoshi value of the output.

`script` is the scriptPubKey for the output (with its length omitted).
`script`s are not required to follow standardness rules: non-standard
scripts such as `OP_RETURN` may be accepted, but the corresponding

> transaction may fail to relay across the network.

Author: The requirement above states:

> The receiving node:
> ...
>  - MAY fail the negotiation if `script` is non-standard

This portion then goes on to say:

>  non-standard scripts such as `OP_RETURN` may be accepted

This seems consistent to me, if not "inverted across the boundary", so to speak.

> `script`s are not required to follow standardness rules: non-standard
> scripts such as `OP_RETURN` may be accepted, but the corresponding
> transaction may fail to relay across the network.

Reviewer 9: The confusing part is:

> The receiving node:
> ...
>  - MAY fail the negotiation if script is non-standard

vs.

> `script`s are not required to follow standardness rules

Something like this would be much clearer:

```suggestion
The receiving node may allow `script`s that do not follow standardness rules (e.g., non-standard
scripts such as `OP_RETURN` may be accepted), but the corresponding
transaction may fail to relay across the network.
```

This message removes an input from the transaction.

1. type: 68 (`tx_remove_input`)
2. data:
    * [`channel_id`:`channel_id`]
    * [`u64`:`serial_id`]

This message removes an output from the transaction.

1. type: 69 (`tx_remove_output`)
2. data:
    * [`channel_id`:`channel_id`]
    * [`u64`:`serial_id`]

#### Requirements

The sending node:
  - MUST NOT send a `tx_remove` with a `serial_id` it did not add
    to the transaction or has already been removed

The receiving node:
  - MUST remove the indicated input or output from the transaction
  - MUST fail the negotiation if:
    - the input or output identified by the `serial_id` was not added by the
      sender
    - the `serial_id` does not correspond to a currently added input (or output)

### The `tx_complete` Message

This message signals the conclusion of a peer's transaction
contributions.

1. type: 70 (`tx_complete`)
2. data:
    * [`channel_id`:`channel_id`]

#### Requirements

The nodes:
  - MUST send this message in succession to conclude this protocol

The receiving node:
  - MUST use the negotiated inputs and outputs to construct a transaction
  - MUST fail the negotiation if:

>     - the peer's total input satoshis is less than their outputs

Author: >  The opener contributes the funding output, which should be allowed to exceed the opener's contributed inputs.

This could be clearer. The peer's contribution to the `funding output` is considered part of their output set for the sake of this check.

The same logic applies for the `splice out` and should, similarly, still be within spec.

    - the peer's paid feerate does not meet or exceed the agreed `feerate`
      (based on the `minimum fee`).
    - if is the *non-initiator*:
      - the *initiator*'s fees do not cover the `common` fields
    - there are more than 252 inputs
    - there are more than 252 outputs

>     - the estimated weight of the tx is greater than 400,000 (`MAX_STANDARD_TX_WEIGHT`)

Reviewer 2: > I think the check should be delayed when all `tx_signatures` have been received ?

Unfortunately you can't, it would be too late by that time. If you're sending `tx_signatures` first, your peer can always choose to publish the transaction, even if their `tx_signatures` has a larger witness than expected. Failing at that point wouldn't be very interesting, you should either wait for the transaction to confirm or double-spend it if you don't see it confirming quickly enough (so we don't need to specifically handle this case).

>     - the estimated weight of the tx is greater than 400,000 (`MAX_STANDARD_TX_WEIGHT`)

Reviewer 3: > If you're sending tx_signatures first, your peer can always choose to publish the transaction, even if their tx_signatures has a larger witness than expected. 

You could sequentialize the `tx_signatures` exchange as this witness weight estimation check is always a possibility for the first-receiver of the `tx_signatures`, though not for the remaining counterparties. I think this has been discussed with previous iterations of the spec, and it sounds the current proposal is “okayish” for now.

I still think it will be a source of DoS for collaborative transaction participants in the future and this stay an interesting problem. You might have margin of improvements like knowing the “worst-policy-size” and based on this ask premium fees for releasing signatures first. Though definitely beyond the scope of this PR.

While the `minimum fee` is calculated and verified at `tx_complete` conclusion,
it is possible for the fee for the exchanged witness data to be underpaid.
It is the responsibility of the sending peer to correctly account for the
required fee.

### The `tx_init_rbf` Message

This message initiates a replacement of the transaction after it's been
completed.

1. type: 72 (`tx_init_rbf`)
2. data:
   * [`channel_id`:`channel_id`]
   * [`u32`:`locktime`]
   * [`u32`:`feerate`]
   * [`tx_init_rbf_tlvs`:`tlvs`]

1. `tlv_stream`: `tx_init_rbf_tlvs`
2. types:
    1. type: 0 (`funding_output_contribution`)
    2. data:
        * [`s64`:`satoshis`]

#### Requirements

The sender:
  - MUST set `feerate` greater than or equal to 25/24 times the `feerate`
    of the previously constructed transaction, rounded down.
  - If it contributes to the transaction's funding output:
    - MUST set `funding_output_contribution`

The recipient:
  - MUST respond either with `tx_abort` or with `tx_ack_rbf`
  - MUST respond with `tx_abort` if:
    - the `feerate` is not greater than or equal to 25/24 times `feerate`
      of the last successfully constructed transaction

>   - MAY send `tx_abort` for any reason

Reviewer 2: Agreed, but I don't think this should be here, or specific to RBF. You should simply check that your peer pays enough fees for their own inputs/outputs at the chosen feerate, and that is already covered by the specification?

>   - MAY send `tx_abort` for any reason

Reviewer 3: > You should simply check that your peer pays enough fees for their own inputs/outputs at the chosen feerate, and that is already covered by the specification?

Yes I think the specification is good enough on this, though I’m worried about mis-implementation issues and out-of-order checks in the way it’s described. LDK code is still under review and implementation, I’ll keep in mind to get right this check.

   * [`tx_ack_rbf_tlvs`:`tlvs`]

1. `tlv_stream`: `tx_ack_rbf_tlvs`
2. types:
    1. type: 0 (`funding_output_contribution`)
    2. data:
        * [`s64`:`satoshis`]

#### Requirements

The sender:
  - If it contributes to the transaction's funding output:
    - MUST set `funding_output_contribution`

The recipient:
  - MUST respond with `tx_abort` or with a `tx_add_input` message,
    restarting the interactive tx collaboration protocol.

#### Rationale

`funding_output_contribution` is the amount of satoshis that this peer
will contribute to the funding output of the transaction, when there is
such an output. Note that it may be different from the contribution
made in the previously completed transaction. If omitted, the sender is
not contributing to the funding output.

It's recommended that a peer, rather than fail the RBF negotiation due to
a large feerate change, instead stop contributing to the funding output,
and decline to participate further in the transaction (by not contributing,
they may obtain incoming liquidity at no cost).

### The `tx_abort` Message

1. type: 74 (`tx_abort`)
2. data:
   * [`channel_id`:`channel_id`]
   * [`u16`:`len`]

>    * [`len*byte`:`data`]

Reviewer 2: It is very useful, it is the same as the `error` or `warning` messages. We do use it extensively in implementations to tell the other peer what went wrong.

#### Requirements

A sending node:
  - MUST NOT have already transmitted `tx_signatures`
  - SHOULD forget the current negotiation and reset their state.

A receiving node:
  - if they have already sent `tx_signatures` to the peer:

>     - MUST NOT forget the channel until any inputs to the negotiated tx

Reviewer 3: I think a recommendation of double-spending the inputs could be introduced as you have the asynchronous case where Alice sends `tx_signatures` at the same time than Bob is sending `tx_abort`. Reception of `tx_abort` on Alice side puts her on a stale.

Reviewer 2: I'm not sure what you mean? Once you've sent `tx_signatures`, you cannot forget this transaction until it has been double-spent. If you receive `tx_abort`, you should free the corresponding utxos to make sure you try double-spending it at some point.

Reviewer 3: > If you receive tx_abort, you should free the corresponding utxos to make sure you try double-spending it at some point.

From my understanding, we’re saying the same thing namely when you receive `tx_abort` you should double-spend the utxos, though formulated differently, so sounds good.

      (based on the `minimum fee`).
    - if is the *non-initiator*:
      - the *initiator*'s fees do not cover the `common` fields
    - there are more than 252 inputs
    - there are more than 252 outputs
    - the estimated weight of the tx is greater than 400,000 (`MAX_STANDARD_TX_WEIGHT`)

#### Rationale

To signal the conclusion of exchange of transaction inputs and outputs.

Upon successful exchange of `tx_complete` messages, both nodes
should construct the transaction and proceed to the next portion of the
protocol. For channel establishment v2, exchanging commitment transactions.

For the `minimum fee` calculation see [BOLT #3](03-transactions.md#calculating-fees-for-collaborative-transaction-construction).

The maximum inputs and outputs are capped at 252. This effectively fixes
the byte size of the input and output counts on the transaction to one (1).

### The `tx_signatures` Message

1. type: 71 (`tx_signatures`)
2. data:
    * [`channel_id`:`channel_id`]
    * [`sha256`:`txid`]
    * [`u16`:`num_witnesses`]
    * [`num_witnesses*witness_stack`:`witnesses`]

1. subtype: `witness_stack`
2. data:

>     * [`u16`:`num_witness_elements`]

Reviewer 2: To be honest I don't really mind changing this if the current encoding is annoying for your implementation, it would be a rather trivial change on the `eclair` side and we have only shipped experimental support for dual funding so far. But I don't know if it would be annoying for `cln`, so I'll let @niftynei decide!

the byte size of the input and output counts on the transaction to one (1).

### The `tx_signatures` Message

1. type: 71 (`tx_signatures`)
2. data:
    * [`channel_id`:`channel_id`]
    * [`sha256`:`txid`]
    * [`u16`:`num_witnesses`]
    * [`num_witnesses*witness`:`witnesses`]

1. subtype: `witness`
2. data:
    * [`u16`:`len`]
    * [`len*byte`:`witness_data`]

#### Requirements

The sending node:
  - if it has the lowest total satoshis contributed, as defined by total
    `tx_add_input` values, or both peers have contributed equal amounts
    but it has the lowest `node_id` (sorted lexicographically):
    - MUST transmit their `tx_signatures` first
  - MUST order the `witnesses` by the `serial_id` of the input they
    correspond to
  - `num_witnesses`s MUST equal the number of inputs they added
  - MUST use the `SIGHASH_ALL` (0x01) flag on each signature

The receiving node:
  - MUST fail the negotiation if:
    - the message contains an empty `witness`
    - the number of `witnesses` does not equal the number of inputs
      added by the sending node
    - the `txid` does not match the txid of the transaction
    - the `witnesses` are non-standard

> 	- a signature uses a flag that is not `SIGHASH_ALL` (0x01)

Author: Fixed. Thanks!

  - SHOULD apply the `witnesses` to the transaction and broadcast it
  - MUST reply with their `tx_signatures` if not already transmitted

#### Rationale

A strict ordering is used to decide which peer sends `tx_signatures` first.
This prevents deadlocks where each peer is waiting for the other peer to
send `tx_signatures`, and enables multiparty tx collaboration.

The `witness_data` is encoded as per bitcoin's wire protocol (a CompactSize number
of elements, with each element a CompactSize length and that many bytes following).

While the `minimum fee` is calculated and verified at `tx_complete` conclusion,
it is possible for the fee for the exchanged witness data to be underpaid.
It is the responsibility of the sending peer to correctly account for the
required fee.

### The `tx_init_rbf` Message

This message initiates a replacement of the transaction after it's been
completed.

1. type: 72 (`tx_init_rbf`)
2. data:
   * [`channel_id`:`channel_id`]
   * [`u32`:`locktime`]
   * [`u32`:`feerate`]
   * [`tx_init_rbf_tlvs`:`tlvs`]

> 1. `tlv_stream`: `tx_init_rbf_tlvs`

Reviewer 2: I missed that earlier, but I believe `tx_init_rbf` and `tx_ack_rbf` should also have the `require_confirmed_inputs` TLVs that `open_channel2` and `accept_channel2` have? We shouldn't implicitly inherit that from the previous pending transaction (especially in the future context of splicing), and even if you keep setting it exactly like you did in `open_channel2` and `accept_channel2`, it doesn't hurt to be explicit?

Reviewer 2: Note that since this is an `even` TLV, we should decide before merging the spec, as this will be incompatible with current readers of the `tx_init_rbf` message. I think we must do it, so we should do it ASAP! Another option though could be to use an odd tlv tag when adding this TLV to `tx_init_rbf` and `splice_init`.

* on-chain funding attempts are relatively infrequent operations
* honest nodes should complete the protocol quickly, reducing the risk of
  conflicts
* failed attempts can simply be retried at no cost

### The `tx_add_output` Message

This message adds a transaction output.

1. type: 67 (`tx_add_output`)
2. data:
    * [`channel_id`:`channel_id`]
    * [`u64`:`serial_id`]
    * [`u64`:`sats`]
    * [`u16`:`scriptlen`]
    * [`scriptlen*byte`:`script`]

#### Requirements

Either node:
  - MAY omit this message

The sending node:
  - MUST add all sent outputs to the transaction
  - if is the *initiator*:
    - MUST send even `serial_id`s
  - if is the *non-initiator*:
    - MUST send odd `serial_id`s

The receiving node:
  - MUST add all received outputs to the transaction
  - MUST accept P2WSH, P2WPKH, P2TR `script`s
  - MAY fail the negotiation if `script` is non-standard
  - MUST fail the negotiation if:
    - the `serial_id` is already included in the transaction
    - the `serial_id` has the wrong parity
    - it has received 4096 `tx_add_output` messages during this negotiation
    - the `sats` amount is less than the `dust_limit`

>     - the `sats` amount is greater than 2,100,000,000,000,000 (`MAX_MONEY`)

Reviewer 11: we should probably not only check that the `sats` amount of this output is less than `MAX_MONEY`, but the sum of all the outputs thus far?

alternatively, we could make the summation check prior to sending `tx_signatures`?

Author: we generally defer full checks until the `tx_complete` step (eg a peer could have an invalid tx inflight at any point prior to sending tx_complete.) these checks happens prior to sending commitment signatures (which is before `tx-sigs`).

on an output by output basis, it's reasonable to bound to "not an outlandish amount of bitcoin". at the `tx_complete` finalization stage, we will also check that they've provided enough inputs to cover whatever amount of outputs they're claiming they can cover.

### The `tx_signatures` Message

1. type: 71 (`tx_signatures`)
2. data:
    * [`channel_id`:`channel_id`]
    * [`sha256`:`txid`]
    * [`u16`:`num_witnesses`]
    * [`num_witnesses*witness`:`witnesses`]

1. subtype: `witness`
2. data:
    * [`u16`:`len`]
    * [`len*byte`:`witness_data`]

#### Requirements

The sending node:
  - if it has the lowest total satoshis contributed, as defined by total
    `tx_add_input` values, or both peers have contributed equal amounts
    but it has the lowest `node_id` (sorted lexicographically):
    - MUST transmit their `tx_signatures` first
  - MUST order the `witnesses` by the `serial_id` of the input they
    correspond to
  - `num_witnesses`s MUST equal the number of inputs they added

The receiving node:
  - MUST fail the negotiation if:
    - the message contains an empty `witness`
    - the number of `witnesses` does not equal the number of inputs
      added by the sending node
    - the `txid` does not match the txid of the transaction
    - the `witnesses` are non-standard
  - SHOULD apply the `witnesses` to the transaction and broadcast it
  - MUST reply with their `tx_signatures` if not already transmitted

Both nodes:

>   - MUST sign the transaction using SIGHASH_ALL

Author: That's reasonable. I've updated the spec to reflect this.

ids.

### `channel_id`, v2

For channels established using the v2 protocol, the `channel_id` is the
SHA256(lesser-revocation-basepoint || greater-revocation-basepoint),
where the lesser and greater is based off the order of the basepoint.

If the peer's revocation basepoint is unknown (e.g. `open_channel2`
and `accept_channel2`), a temporary `channel_id` should be found by
using a zeroed out basepoint for the unknown peer.

#### Rationale
These values must be remembered by both peers for correct operation anyway.
They're known from the first exchange of messages, obviating the need for
a `temporary_channel_id`. Finally, by mixing information from both sides,
they avoid `channel_id` collisions.

## Interactive Transaction Construction

Interactive transaction construction allows two peers to collaboratively
build a transaction for broadcast.  This protocol is the foundation
for dual-funded channels establishment (v2).

### Set-Up and Vocabulary

There are two parties to a transaction construction: an *initiator*
and a *non-initiator*.
 The *initiator* is the peer which initiates the protocol, e.g.
for channel establishment v2 the *initiator* would be the peer which

> sends `open_channel2`; for a close, the *initiator* sends `shutdown`.

Reviewer 8: I don't think the characterization works for a close, because both parties send `shutdown`, and they can do so concurrently.

MAY omit this message.

`serial_id` is a randomly chosen number which uniquely identifies this input.
Inputs in the constructed transaction are sorted by `serial_id`.

`prevtx` is the serialized transaction that contains the output
this input spends. Used to verify that the input is non-malleable.

`prevtx_vout` is the index of the output being spent.

`sequence` is the sequence number of this input.  Must be less than
4294967294 (0xFFFFFFFE). See BIP125.

`script_sig` is the scriptSig for the input, with length omitted.
The `script_sig` for non-P2SH-wrapped inputs will be an empty byte array.

### The `tx_add_output` Message

This message adds a transaction output.

1. type: 67 (`tx_add_output`)
2. data:
    * [`channel_id`:`channel_id`]
    * [`u64`:`serial_id`]
    * [`u64`:`sats`]
    * [`u16`:`scriptlen`]
    * [`scriptlen*byte`:`script`]

#### Requirements

Either node:
  - MAY omit this message

The sending node:
  - MUST add all sent outputs to the transaction
  - if is the *initiator*:
    - MUST send even `serial_id`s
  - if is the *non-initiator*:
    - MUST send odd `serial_id`s

The receiving node:
  - MUST add the specified output to the transaction

>   - MUST accept P2WSH, P2WPKH, P2PKH, P2TR `script`s

Reviewer 3: omission of P2SH yet P2PKH ?

Reviewer 2: Is there a good reason to accept P2PKH? Shouldn't we push for native segwit only?

Reviewer 2: Is there a good reason to accept P2PKH? Shouldn't we push for native segwit only?

  - MAY fail the negotiation if `script` is non-standard
  - MUST fail the negotiation if:
    - the `serial_id` is already included in the transaction
    - the `serial_id` has the wrong parity
    - it has received 4096 `tx_add_output` messages during this negotiation

>     - the `sats` amount is less than the `dust_limit`

Reviewer 3: maybe link https://github.com/lightning/bolts/blob/master/03-transactions.md#dust-limits

    * [`u16`:`num_witnesses`]
    * [`num_witnesses*witness_stack`:`witness_stack`]

1. subtype: `witness_stack`
2. data:
    * [`u16`:`num_input_witness`]
    * [`num_input_witness*witness_element`:`witness_element`]

1. subtype: `witness_element`
2. data:
    * [`u16`:`len`]
    * [`len*byte`:`witness`]

#### Requirements
The sending node:
  - if it has the lowest total satoshis contributed, as defined
    by total `tx_add_input` values:
    - MUST transmit their `tx_signatures` before the peer
  - MUST order the `witness_stack`s by the `serial_id` of the input they
    correspond to
  - number of `witness_stack`s MUST equal the number of inputs they added

The receiving node:
  - MUST fail the negotiation if:
    - if they have not sent a `tx_signatures` message and they are the
      peer with the lowest total satoshis contributed, as defined by
      `tx_add_input` values
    - the message contains an empty `witness_stack`
    - the number of `witness_stack`s does not equal the number of inputs
      added by the sending node
    - the `txid` does not match the txid of the transaction
  - SHOULD apply the `witness`es to the transaction and broadcast it

>   - MUST reply with their `tx_signatures` if not already transmitted

Reviewer 3: "MAY fail the negotiation if the witnesses are non-standard" ?

    the funding transaction

#### Requirements
The sending node:
  - MUST verify it has received a valid commitment signature from its peer
  - MUST remember the details of this funding transaction
  - if it has NOT received a valid `commitment_signed` message:
    - MUST NOT send a `tx_signatures` message

The receiving node:
  - if the received `witness_stack` weight results in the peer's
    paid feerate falling below the *opener*'s feerate for the funding
    transaction:
    - SHOULD broadcast their commitment transaction, closing the channel.
  - SHOULD apply `witness`es to the funding transaction and broadcast it
  - if has already sent or received a `funding_locked` message for this
    channel:
    - MUST ignore this message

#### Rationale
A peer sends their `tx_signatures` as soon as they have received a valid
`commitments_signed` message.

The channel should be preemptively closed in the case where a peer provides
valid witness data that causes their paid feerate to fall beneath the
`open_channel2.funding_feerate_perkw` rate. This penalizes the peer
for underpayment of fees.

### The `init_rbf` Message

> This message initiates a replacement of a broadcast funding transaction.

Author: > Should we add a note somewhere about saying CPFP can be considered if a change output is available on the interactive transaction ?

Astute bitcoiners will know this is implicitly true, nothing about this protocol changes the ground-rules for CPFP eligibility -- not sure it's worth calling out explicitly.

> This message initiates a replacement of a broadcast funding transaction.

Author: > at the benefit of non-interactivity.

Worth pointing out that 'non-interactivity' or reduction in interactivity is a noble goal for most cryptographic protocols, however the establishment of a lightning channel is an on-going, interactive relationship so shouldn't weigh so heavily in this protocol as it might elsewhere.

#### Requirements

The sender:
  - MUST have sent `open_channel2`
  - MUST set `fee_step` greater than zero and greater than any prior `fee_step`
  - MUST NOT have sent or received a `funding_locked` message.

The recipient:
  - MUST respond with either an error or an `ack_rbf` message.
  - MUST fail the negotiation if:
    - the `fee_step` is not greater than the last successfully negotiated
      `init_rbf` attempt or `one` if no prior successful `init_rbf` has
       been received
    - they have already sent or received `funding_locked`
  - MAY fail the negotiation for any reason

#### Rationale
`fee_step` is an integer value, which specifies the `feerate` for this
funding transaction, as a rate of increase above the `open_channel2`.
`funding_feerate_perkw`.

The effective `funding_feerate_perkw` for this RBF attempt
if calculated as 1.25^`fee_step` * `funding_feerate_perkw`.
E.g. if `feerate_per_kw_funding` is 512 and the `fee_step` is 1,
the effective `feerate` for this RBF attempt is 512 + 512 / 4 or 640 sat/kw.
A `fee_step` 2 would be `1.25^2 * 512` (or 640 + 640 / 4), 800 sat/kw.

If a valid `funding_locked` message is received in the middle of an

> RBF attempt, the attempt MUST be abandoned.

Author: Ideally yes, but in practice it's impossible to pre-commit to a feerate that is guaranteed to surpass the last tx's absolute fee, given that the input set can be reconfigured.

Open to suggestions on how to enforce this however.

> RBF attempt, the attempt MUST be abandoned.

Author: As currently implemented, in some cases you have to repeatedly RBF (and increment the feerate) until you hit upon a tx that works -- really not ideal but it works.

### The `ack_rbf` Message

1. type: 73 (`ack_rbf`)
2. data:
    * [`channel_id`:`channel_id`]
    * [`u64`:`funding_satoshis`]

#### Requirements

The sender:
  - MUST NOT have sent or received a `funding_locked` message

The recipient:

>   - MUST either fail the negotiation or transmit a `tx_add_input` message

Author: The 'interactive tx construction' flow should be restarted. Attempted to make this clear in the message diagram, above.

```
        +-------+                              +-------+
        |       |--(1)--- open_channel2  ----->|       |
        |       |<-(2)--- accept_channel2 -----|       |
        |       |                              |       |
    --->|       |      <tx collaboration>      |       |
    |   |       |                              |       |
    |   |       |--(3)--  commitment_signed -->|       |
    |   |       |<-(4)--  commitment_signed ---|       |
    |   |   A   |                              |   B   |
    |   |       |<-(5)--  tx_signatures -------|       |
    |   |       |--(6)--  tx_signatures ------>|       |
    |   |       |                              |       |
    |   |       |--(a)--- init_rbf ----------->|       |
    ----|       |<-(b)--- ack_rbf  ------------|       |
        |       |                              |       |
        |       |--(7)--- funding_locked ----->|       |
        |       |<-(8)--- funding_locked ------|       |
        +-------+  
````

#### Rationale
`funding_satoshis` is the amount of satoshis that this peer will
contribute to the funding output.  Note that this may be different than
the amount transmitted in either `accept_channel2`.`funding_satoshis` or
any previous `ack_rbf`.`funding_satoshis`, as the amount the *accepter*
wishes to commit to the funding output may change.

It's recommended that a peer, rather than fail the RBF negotiation due to
a large feerate change, instead sets their `funding_satoshis` to zero,

> and decline to participate further in the channel funding.

Author: > Should init_rbf/ack_rbf be transaction-level messages and not channel-level ones ?

I'm not sure I follow what you mean by this -- as described they re-initiate the tx-construction phase of the protocol.

   * [`u32`:`funding_feerate_perkw`]
   * [`u32`:`commitment_feerate_perkw`]
   * [`u64`:`funding_satoshis`]
   * [`u64`:`dust_limit_satoshis`]
   * [`u64`:`max_htlc_value_in_flight_msat`]
   * [`u64`:`htlc_minimum_msat`]
   * [`u16`:`to_self_delay`]
   * [`u16`:`max_accepted_htlcs`]
   * [`u32`:`locktime`]
   * [`point`:`funding_pubkey`]
   * [`point`:`revocation_basepoint`]
   * [`point`:`payment_basepoint`]
   * [`point`:`delayed_payment_basepoint`]
   * [`point`:`htlc_basepoint`]
   * [`point`:`first_per_commitment_point`]
   * [`byte`:`channel_flags`]
   * [`opening_tlvs`:`tlvs`]

1. `tlv_stream`: `opening_tlvs`
2. types:
   1. type: 1 (`option_upfront_shutdown_script`)
   2. data:
       * [`u16`:`shutdown_len`]
       * [`shutdown_len*byte`:`shutdown_scriptpubkey`]

Rationale and Requirements are the same as for [`open_channel`](#the-open_channel-message), with the following additions:

#### Requirements:

If nodes have negotiated `option_dual_fund`:
  - the opening node:
    - MUST NOT send `open_channel`

The sending node:
  - MUST set `funding_feerate_perkw` to the feerate for this transaction
  - MUST ensure `temporary_channel_id` is unique from any

>     other channel ID with the same peer.

Author: it is!

    - the `locktime` is unacceptable
    - the `funding_feerate_per_kw` is unacceptable

#### Rationale
`channel_id` for the `open_channel2` MUST be derived using a zero-d out
basepoint for the peer's revocation basepoint.  This allows the peer to
return channel-assignable errors before the *accepter*'s revocation
basepoint is known.

`funding_feerate_perkw` indicates the fee rate that the opening node will
pay for the funding transaction in satoshi per 1000-weight, as described
in [BOLT-3, Appendix F](03-transactions.md#appendix-f-dual-funded-transaction-test-vectors).

`locktime` is the locktime for the funding transaction.

The receiving node, if the `locktime` or `feerate_funding_perkw` is considered
out of an acceptable range, may fail the negotiation. However, it is
recommended that the *accepter* permits the channel open to proceed
without their participation in the channel's funding.

Note that `open_channel`'s `channel_reserve_satoshi` has been omitted.
Instead, the channel reserve is fixed at 1% of the total channel balance
(`open_channel2`.`funding_satoshis` + `accept_channel2`.`funding_satoshis`)
rounded down to the nearest whole satoshi or the `dust_limit_satoshis`,

> whichever is greater.

Reviewer 2: Can you elaborate on why the explicit reserve was removed?
We find it useful to be able to explicitly specify it to allow setting it to `0` for mobile wallet peers (only on the wallet side), where having a channel reserve makes understanding the balance very hard for users (which means our node takes a risk that the mobile wallet peer may misbehave, but since they pay us fees it's a trade-off we choose to make for a better UX).

Author: We removed it so as to reduce the number of things that are communicated between peers. This assumed that every channel needed/would have a reserve requirement, and we picked a 'sane, arbitrary' default for it.

If this is a useful feature to be able to negotiate, we can certainly add it back.

Is being able to negotiate the reserve amount useful, or should we simply turn it into a flag that signals to use/not use the reserve.

Reviewer 3: I think here I would advocate to keep the reserve negotiation. Existence of the reserve could be a data point that the game-theory backing up the LN-penalty mechanism is empirically sounds and one of the reason we see few attacks by broadcasting a revoked state today on the network.

On the otherside, I think it make sense to set up to low amounts (even `0`) where there are other out-of-band enforcement mechanism of the channel state correctness (e.g business brand of a LSP).

If a node operator is interested to benefit from the advantages of `channelv2` and would like to enforce a significant channel reserve on its peers, it could make sense to make the reserve dynamic and subject to negotiation.

Reviewer 2: I was thinking the same @ariard, but then I realized that:

- 1% is a good default since most channels won't care
- we can introduce (later) a tlv field for optional reserve negotiation: that's what we're going to do for Phoenix

So I think it's good that the default case doesn't contain a configurable reserve, since it optimizes the most common scenario while allowing configuration via a future tlv field.

Reviewer 3: Optional reserve negotiation SGTM!

Reviewer 12: > 1% is a good default since most channels won't care about configuring it properly

If every wallet is going to discard the 1% reserve requirement then it should be mentioned in here that it's not a requirement. Unless you explicitly want everyone to go against the BOLT spec (which what is the purpose of a spec?), then it should be clarified.

Reviewer 2: I disagree, wallets are a special case of a lightning node. The spec should decide what the default is, and special cases should override with extensions. It should be a wallet extension feature to use 0-reserve (which I proposed [here](https://github.com/lightning/blips/pull/30)).

    * [`point`:`first_per_commitment_point`]
    * [`accept_tlvs`:`tlvs`]

1. `tlv_stream`: `accept_tlvs`
2. types:
   1. type: 1 (`option_upfront_shutdown_script`)
   2. data:
       * [`u16`:`shutdown_len`]
       * [`shutdown_len*byte`:`shutdown_scriptpubkey`]

Rationale and Requirements are the same as listed above,
for [`accept_channel`](#the-accept_channel-message) with the following
additions.

#### Requirements:

The accepting node:
    - MAY respond with a `funding_satoshis` value of zero.

#### Rationale

The `funding_satoshis` is the amount of bitcoin in satoshis
the *accepter* will be contributing to the channel's funding transaction.

Note that `accept_channel`'s `channel_reserve_satoshi` has been omitted.
Instead, the channel reserve is fixed at 1% of the total channel balance
(`open_channel2`.`funding_satoshis` + `accept_channel2`.`funding_satoshis`)
rounded down to the nearest whole satoshi or the `dust_limit_satoshis`,
whichever is greater.

`channel_id` for the `accept_channel2` MUST be derived using a zero-d out
basepoint for the peer's revocation basepoint.  This allows the peer to
return channel-assignable errors before the *accepter*'s revocation

> basepoint is known.

Reviewer 2: Is that a bad copy-paste? We received the peer's revocation basepoint in `open_channel2` so we should be able to use the final `channel_id` here, can't we?

Author: Originally, I thought the same thing and in fact the v0.10.0 implementation of c-lightning did exactly this but @rustyrussell pointed out that it's exceedingly bad networking practice to have an id rely on data contained within the message, as that builds a dependency between knowing who/where to route the message to and the knowledge of how to unpack the message (tho your router needs to know to pull the channel_id out... so hm).

Reviewer 2: That's an interesting argument! But unless I'm missing something, it doesn't hold at all in this case...because you still have `zerod_channel_id` depend on the `revocation_basepoint` inside the message in both `open_channel2` and `accept_channel2`, don't you?

If I read this correctly, `open_channel2` still uses the opener's `revocation_basepoint` inside its own message to compute its `zerod_channel_id`. And similarly, the accepter uses its own `revocation_basepoint` from its `accept_channel2` message to compute its `zerod_channel_id`. So both messages still conflict with @rustyrussell's rule!

What I'm suggesting is that `accept_channel2` actually incorporate information from `open_channel2` (a different message) to provide the final `channel_id` since it can! Am I missing the point completely or not?

Reviewer 2: And thinking about this more, I believe we *must* use the final `channel_id` (that combines both `revocation_basepoint`s) in `accept_channel2`.

Imagine you're trying to open two distinct channels at the same time and sent two `open_channel2` to a peer (ok, that's a bit contrived, there may be no reason to do that but still).
They respond with two `accept_channel2`, but which one is for which channel?
You can only know if `accept_channel2` contains a reference to *something* in the corresponding `open_channel2`, which is achieved by using a `channel_id` that includes `open_channel2.revocation_basepoint`.

Reviewer 2: Note for future reviewers: my last two comments were based on a misunderstanding. The `zerod_channel_id` sent in `accept_channel2` is the same value that was received in `open_channel2`. Which is perfectly fine and addresses my comments.

### Funding Composition
Funding composition for channel establishment v2 makes use of the
[Interactive Transaction Construction](#interactive-transaction-construction)
protocol, with the following additional caveats.

#### The `tx_add_input` Message

No additional caveats or requirements.

#### The `tx_add_output` Message
##### Requirements
The sending node:
  - if is the *opener*:
    - MUST send at least one `tx_add_output`,  which contains the

>       channel's funding output

Reviewer 2: Maybe dumb question, but how does the `accepter` know which output is the funding output?
Don't we need to exchange its `serial_id` beforehand (for example it could be in `open_channel2`)?

Author: You can figure it out based on the funding_pubkeys that are exchanged in `open_channel2` and `accept_channel2`.

Author: You'd want to independently verify this anyway, instead of relying on your peer to tell you they did it.

Reviewer 2: Good point, that was a dumb question! Sorry I'm new to this lightning thing :laughing:

##### Rationale
The channel funding output must be added by the *opener*, who pays its fees.

#### The `tx_complete` Message

Upon receipt of consecutive `tx_complete`s, the receiving node:
  - if is the *accepter*:
    - MUST fail the negotiation if:
      - no funding output was received
      - the value of the funding output is not equal to the sum of
        `open_channel2`.`funding_satoshis` and `accept_channel2`.
        `funding_satoshis`
      - the value of the funding output is less than the `dust_limit`

>  - if is an RBF attempt:

Reviewer 2: nit:
```suggestion
 - if it is an RBF attempt:
```

    - MUST fail the negotiation if:
      - the transaction's total fees is less than the last
        successfully negotiated transaction's fees
      - the transaction does not share a common input with all previous

>         funding transactions

Reviewer 2: I think we need to be more restrictive than that, we also cannot add new unconfirmed inputs (BIP 125 rule 2).
Or maybe should we say "the transaction follows the usual BIP 125 replacement rules" or something like it?

Author: Ideally yes, we'd specify that they can't break any of the BIP125 rules, including the "no new unconfirmed inputs". However, we can't really verify that without attempting to broadcast it to our local mempool. We *can* however verify that they're executing a doublespend (and in fact this is more important to verify than "no new unconfirmed inputs"), as failing to adhere to it might lead to *two* channel funding txs being included in the chain.

I tried to adhere to everything we absolutely needed to verify/confirm and leave out the things that we can't really (but if we could we should), such as the "no new unconfirmed inputs" rule. 

An RBF might fail to broadcast for a few reasons, and it's expected that they will.

Author: I could add a note that the *sender* shouldn't send any new unconfirrmed inputs, but there's no enforcement on the receiver side really.

Reviewer 2: Note that if you're using `bitcoind`, you actually can perform the checks without broadcasting with the `testmempoolaccept` RPC (that's probably what I'd use when implementing).

> I could add a note that the sender shouldn't send any new unconfirrmed inputs, but there's no enforcement on the receiver side really. What do you think?

That sounds good enough!

Reviewer 2: That's true.

  - MUST transmit a [`tx_signatures` message](#the-tx_signatures-message) with their signatures for
    the funding transaction

#### Requirements
The sending node:
  - MUST verify it has received a valid commitment signature from its peer
  - MUST remember the details of this funding transaction
  - if it has NOT received a valid `commitment_signed` message:
    - MUST NOT send a `tx_signatures` message

The receiving node:
  - if the received `witness_stack` weight results in the peer's
    paid feerate falling below the *opener*'s feerate for the funding
    transaction:
    - SHOULD broadcast their commitment transaction, closing the channel.
  - SHOULD apply `witness`es to the funding transaction and broadcast it
  - if has already sent or received a `funding_locked` message for this
    channel:
    - MUST ignore this message

#### Rationale
A peer sends their `tx_signatures` as soon as they have received a valid
`commitments_signed` message.

The channel should be preemptively closed in the case where a peer provides
valid witness data that causes their paid feerate to fall beneath the
`open_channel2.funding_feerate_perkw` rate. This penalizes the peer

> for underpayment of fees.

Reviewer 2: There is duplication with the rationale in the `tx_signatures` paragraph, maybe we should keep only one of the two (or just link to it here)?

Author: Cut this down to just here based on an earlier comment.

### The `init_rbf` Message

This message initiates a replacement of a broadcast funding transaction.

1. type: 72 (`init_rbf`)
2. data:
   * [`channel_id`:`channel_id`]
   * [`u64`:`funding_satoshis`]
   * [`u32`:`locktime`]
   * [`u32`:`funding_feerate_perkw`]

#### Requirements

The sender:
  - MUST have sent `open_channel2`
  - MUST set `funding_feerate_perkw` greater than or equal to
    65/64 times the last sent `funding_feerate_perkw` rounded down.
  - MUST NOT have sent or received a `funding_locked` message.

The recipient:

>   - MUST respond with either an error or an `ack_rbf` message.

Reviewer 2: I think we shouldn't use the `error` message here, because at that point it conflicts with the requirement to abandon the channel if an `error` is received...
An explicit message to `reject_rbf` feels cleaner, as it only rejects that rbf attempt without impacting the rest of the channel flow.

Author: I think I mean `warning`; basically "fail the negotiation". In fact, at any point in time the peer can fail the RBF negotiation (send a warning and hang up). The peer should reconnect and re-try. Or reconnect and allow the channel open to proceed (assuming the original funding tx gets mined).

  - MUST fail the negotiation if:
    - the `funding_feerate_perkw` is not greater than 65/64 times
      `funding_feerate_perkw` of the last successfully negotiated
      open attempt
    - they have already sent or received `funding_locked`
  - MAY fail the negotiation for any reason

#### Rationale
`funding_feerate_perkw` is the feerate this funding transaction

> will pay. It must be at least 1/64 greater than the last

Reviewer 2: Where does 1/64 come from? It will not necessarily make the BIP 125 checks pass (but that could be the responsibility of the peers, to use some greater if needed to pass BIP 125 rules)...

Author: It's an arbitrary minimum bar to ensure that your peer doesn't endlessly RBF the channel. 1/64 is 1.5% greater than the last (as a floor); it takes ~45 rounds at most to double the feerate.

We could pick a different number; 1/8th takes about 6 rounds to double.

Note that there's now way to enforce a feerate that guarantees you'll pass the BIP125 checks, as the feerate is communicated prior to the input/outputs for the replacement tx.

Reviewer 2: Got it, I think that I'll use a bigger incremental bump than that in eclair, but I don't mind having this value in the spec (you're right that we just need "some" value).

        successfully negotiated transaction's fees
      - the transaction does not share a common input with all previous
        funding transactions

### The `commitment_signed` Message

This message is exchanged by both peers. It contains the signatures for
the first commitment transaction.

Rationale and Requirements are the same as listed below,
for [`commitment_signed`](#commiting-updates-so-far-commitment_signed) with the following additions.

#### Requirements

The sending node:
  - MUST send zero HTLCs.

The receiving node:
  - if the message has one or more HTLCs:
    - MUST fail the negotiation
  - if it has not already transmitted its `commitment_signed`:
    - MUST send `commitment_signed`
  - Otherwise:
    - MUST send `tx_signatures`

#### Rationale

The first commitment transaction has no HTLCs.

### Sharing funding signatures: `tx_signatures`

After a valid `commitment_signature` has been received
from the peer and a `commitment_signature` has been sent, a peer:
  - MUST transmit a [`tx_signatures` message](#the-tx_signatures-message) with their signatures for
    the funding transaction

#### Requirements

> The sending node:

Reviewer 2: We should add a requirement to retransmit signatures on reconnection:

```
  - MUST retransmit `tx_signatures` when reconnecting
```

Otherwise we may stumble upon the following dead-lock:

- Alice and Bob have exchanged everything up to `commit_sig`
- Alice goes first and sends her `tx_signatures`
- But Bob disconnects without receiving them
- On reconnection:
  - Alice:
    - is waiting for Bob's `tx_signatures` to be able to publish the funding tx
    - cannot forget the channel because she doesn't know whether Bob has received her signatures or not
  - Bob:
    - has forgotten the channel because he never received Alice's `tx_signatures`
    - cannot prove it to Alice

If peers retransmit their latest `tx_signatures` on reconnection, we get rid of this deadlock.
Does that make sense?

Also, it may be time to squash and rebase :)

Reviewer 9: Thinking about this more, perhaps retransmitting `tx_signatures` is unnecessary complexity.

Alice can handle this situation in the same way she would if Bob intentionally withheld `tx_signatures` (i.e. monitor the blockchain for the pending channel until an input is double spent).  We need to handle this scenario anyway, so retransmitting `tx_signatures` doesn't seem to simplify implementation at all or prevent any griefing attack.

Reviewer 2: > Thinking about this more, perhaps retransmitting tx_signatures is unnecessary complexity.

Why? It really isn't complex at all: just send `tx_signatures` again when reconnecting if you know that channel. At that point you don't have to care about the ordering of signers because you should have forgotten the channel on disconnection if you had not already sent your `tx_signatures` (similarly to how in channel open v1, the accepting node forgets the channel is it gets disconnected before sending `funding_signed`).

It's much easier to explicitly receive your peer's `tx_signatures` (and thus keep a fully signed version of the transaction in your state) than having to monitor the blockchain to extract it IMO.

Reviewer 9: I think I was a bit confused by the Alice/Bob scenario given, since retransmitting doesn't change the outcome there.  Bob forgot the channel.  Alice will resend her `tx_signatures`, and Bob will ignore them or send an error.  So Alice still needs to monitor the blockchain until an input is double spent.

But the reverse scenario is where retransmitting would help.  Bob gets Alice's `tx_signatures` and sends his `tx_signatures`, but Alice disconnects without receiving them.  When Alice restarts, she and Bob will re-exchange signatures, allowing her to finalize the funding tx and broadcast it.

Alternatively, we could go without the re-exchange, and Alice could notice the funding tx was broadcast by Bob.  We need to watch for the funding tx (or any input double-spend) anyway if Bob refuses to send his `tx_signatures`.

> It's much easier to explicitly receive your peer's tx_signatures (and thus keep a fully signed version of the transaction in your state) than having to monitor the blockchain to extract it IMO.

Since we need to handle a missing `tx_signatures` anyway, the code already needs to be implemented.  I think it would be trivial to call it in one new place (on restart).

The second `tx_signatures` is always redundant in the protocol, since the receiver of the first `tx_signatures` can broadcast the funding tx.  So perhaps the root question is: what benefit do we get from sending `tx_signatures` both ways?  And is it worth ensuring we do this in the case of an untimely disconnect?

Reviewer 2: > Since we need to handle a missing tx_signatures anyway, the code already needs to be implemented.

What code are you referring to? We actually don't need to implement monitoring the mempool to get the fully signed transaction: when we don't have the remote signatures we simply don't broadcast the funding transaction (and rely on our peer to do it), but the protocol state machine isn't impacted. Eclair never fetches the funding tx from the mempool when it is missing the remote signatures, and there's no reason to create a dependency on the mempool here.

> The second tx_signatures is always redundant in the protocol, since the receiver of the first tx_signatures can broadcast the funding tx. 

But that's not the end of the story. Consider the case of a mobile wallet and an LSP, where the LSP sends their `tx_signatures` first and doesn't receive the mobile user's `tx_signatures` the first time (because the connection dropped or whatever). The mobile wallet user could regularly go offline for a while, and doesn't necessarily have a very reliable way of broadcasting on-chain txs: since they're offline most of the time, their wallet won't automatically rebroadcast the funding transaction if it gets evicted from the mempool (mobile wallet users simply don't watch the mempool). It is really beneficial for that mobile wallet user to let the LSP regularly rebroadcast the funding transaction to get it confirmed, especially if the mempool is empty at some point while the mobile wallet user is offline. That is only possible if the LSP gets the remote `tx_signatures`, which requires re-sending them.

Reviewer 9: Ack.  The LSP/mobille case is something I wasn't thinking about.  You've convinced me that adding retransmission is worth it.

    ----|       |<-(b)--- tx_ack_rbf ----------|       |
        |       |                              |       |
        |       |--(7)--- channel_ready  ----->|       |
        |       |<-(8)--- channel_ready  ------|       |
        +-------+                              +-------+

        - where node A is *opener*/*initiator* and node B is
          *accepter*/*non-initiator*

### The `open_channel2` Message

This message initiates the v2 channel establishment workflow.

1. type: 64 (`open_channel2`)
2. data:
   * [`chain_hash`:`chain_hash`]
   * [`channel_id`:`zerod_channel_id`]
   * [`u32`:`funding_feerate_perkw`]
   * [`u32`:`commitment_feerate_perkw`]
   * [`u64`:`funding_satoshis`]
   * [`u64`:`dust_limit_satoshis`]
   * [`u64`:`max_htlc_value_in_flight_msat`]
   * [`u64`:`htlc_minimum_msat`]
   * [`u16`:`to_self_delay`]
   * [`u16`:`max_accepted_htlcs`]
   * [`u32`:`locktime`]
   * [`point`:`funding_pubkey`]
   * [`point`:`revocation_basepoint`]
   * [`point`:`payment_basepoint`]
   * [`point`:`delayed_payment_basepoint`]
   * [`point`:`htlc_basepoint`]
   * [`point`:`first_per_commitment_point`]
   * [`byte`:`channel_flags`]
   * [`opening_tlvs`:`tlvs`]

> 1. `tlv_stream`: `opening_tlvs`

Reviewer 2: Can we also add an empty tlv to both `open_channel2` and `accept_channel2` that signals that a node will only accept confirmed inputs? Something like (I'm not sure how to define an empty tlv without breaking your spec parsing):

```
1. type: 2 (`confirmed_inputs_only`)
2. data:
    * []
```

Only accepting confirmed inputs ensures that the funding tx doesn't end up over-paying fees to make up for low-feerate unconfirmed ancestors of your counterparty. This can be quite useful for LSPs who end up opening a lot of dual-funded channels with wallet users and want to make sure they don't get "robbed" on on-chain fees.

whichever is greater.

Note that `push_msat` has been omitted.

### The `accept_channel2` Message

This message contains information about a node and indicates its
acceptance of the new channel.

1. type: 65 (`accept_channel2`)
2. data:
    * [`channel_id`:`zerod_channel_id`]
    * [`u64`:`funding_satoshis`]
    * [`u64`:`dust_limit_satoshis`]
    * [`u64`:`max_htlc_value_in_flight_msat`]
    * [`u64`:`htlc_minimum_msat`]
    * [`u32`:`minimum_depth`]
    * [`u16`:`to_self_delay`]
    * [`u16`:`max_accepted_htlcs`]
    * [`point`:`funding_pubkey`]
    * [`point`:`revocation_basepoint`]
    * [`point`:`payment_basepoint`]
    * [`point`:`delayed_payment_basepoint`]
    * [`point`:`htlc_basepoint`]
    * [`point`:`first_per_commitment_point`]
    * [`accept_tlvs`:`tlvs`]

1. `tlv_stream`: `accept_tlvs`
2. types:
   1. type: 0 (`upfront_shutdown_script`)
   2. data:
       * [`...*byte`:`shutdown_scriptpubkey`]
   1. type: 1 (`channel_type`)
   2. data:
        * [`...*byte`:`type`]

Rationale and Requirements are the same as listed above,
for [`accept_channel`](#the-accept_channel-message) with the following
additions.

#### Requirements

The accepting node:

>     - MAY respond with a `funding_satoshis` value of zero.

Reviewer 2: We just finalize the v2 protocol, with an empty `tx_signatures` from the accepter.

> this would conflict with the requirement that tx_signatures has a nonempty witness_stack

Can you point me to which requirement that is?

> Also, could the accepter still add inputs/outputs to the funding tx without contributing to the funding output?

Yes, that should be accepted!

>     - MAY respond with a `funding_satoshis` value of zero.

Reviewer 9: > > this would conflict with the requirement that tx_signatures has a nonempty witness_stack
> 
> Can you point me to which requirement that is? Note the distinction between `witnesses` and `witness_stack`: you are allowed to send 0 witnesses, but if you send a witness, then its `witness_stack` must be non-empty.

Ah, the namings are ambiguous (I mentioned this is a previous review).  From "The `tx_signatures` Message":

```
1. type: 71 (`tx_signatures`)
2. data:
    * [`channel_id`:`channel_id`]
    * [`sha256`:`txid`]
    * [`u16`:`num_witnesses`]
    * [`num_witnesses*witness_stack`:`witness_stack`]

1. subtype: `witness_stack`
2. data:
    * [`u16`:`num_input_witness`]
    * [`num_input_witness*witness_element`:`witness_element`]
```

Notice how both the subtype *and* the element in `tx_signatures` are called `witness_stack`?  This makes it unclear what the requirement below refers to:

```
The receiving node:
  - MUST fail the negotiation if:
  ...
    - the message contains an empty `witness_stack`
```

Should we fail if the element `witness_stack` is empty (i.e. `num_witnesses` is 0), or if the subtype `witness_stack` is empty (i.e. `num_input_witness` is 0).

#### Rationale

The `funding_satoshis` is the amount of bitcoin in satoshis
the *accepter* will be contributing to the channel's funding transaction.

Note that `accept_channel`'s `channel_reserve_satoshi` has been omitted.
Instead, the channel reserve is fixed at 1% of the total channel balance
(`open_channel2`.`funding_satoshis` + `accept_channel2`.`funding_satoshis`)
rounded down to the nearest whole satoshi or the `dust_limit_satoshis`,
whichever is greater.

`zerod_channel_id` for the `accept_channel2` MUST be derived using a zero-d out
basepoint for the peer's revocation basepoint.  This allows the peer to
return channel-assignable errors before the *accepter*'s revocation

> basepoint is known.

Reviewer 2: The `zerod_channel_id` is always using:

- the real revocation basepoint of the initiator (because they start)
- a bunch of zeroes instead of the non-initiator basepoint

This ensures the value is the same in `open_channel2` and `accept_channel2` which lets the initiator match the `accept_channel2` to the `open_channel2` it sent (in case it's opening multiple channels).

Basically the non-initiator should just take the `temporary_channel_id` from `open_channel2` and use the same value in its `accept_channel2` response.

> basepoint is known.

Reviewer 9: Thanks, @t-bast -- this makes the most sense.

It would be good to clarify the spec.  As a first-time reader, it was confusing to me.

from the peer and a `commitment_signature` has been sent, a peer:
  - MUST transmit a [`tx_signatures` message](#the-tx_signatures-message) with their signatures for
    the funding transaction

#### Requirements

The sending node:
  - MUST verify it has received a valid commitment signature from its peer
  - MUST remember the details of this funding transaction
  - if it has NOT received a valid `commitment_signed` message:
    - MUST NOT send a `tx_signatures` message

The receiving node:
  - if the `witness_stack` weight lowers the effective `feerate`
    below the the *opener*'s feerate for the funding transaction:
    - SHOULD broadcast their commitment transaction, closing the channel.
  - SHOULD apply `witnesses` to the funding transaction and broadcast it
  - if has already sent or received a `channel_ready` message for this
    channel:
    - MUST ignore this message

#### Rationale

A peer sends their `tx_signatures` as soon as they have received a valid
`commitment_signed` message.

The channel should be preemptively closed in the case where a peer provides
valid witness data that causes their paid feerate to fall beneath the

> `open_channel2.funding_feerate_perkw` rate. This penalizes the peer

Reviewer 3: I'm not sure this punishment is effectively sound, in the sense if you already send your witnesses then you receive the counterparty ones and check that the witness data causes their paid feerate to fall beneath the `funding_feerate_perkw` closing the channel by broadcasting the commitment transaction might penalize yourself twice 1) your fees contributions contributing to counterparties utxo spend and 2) your CPFP adjusting the commitment transaction feerate in a post-anchor world ?

I'm not sure there is a fair and clear solution to this asymmetry, unless some out-of-band fee to cover the party encumbering the risk. I think for now we can live with it and if this utxo-spending free-ridding becomes an effective issue, apply some mitigation.

Reviewer 2: It depends on what we mean by "preemptively closed" here. I agree with you that if the funding transaction is already broadcast, it's dumb to penalize yourself twice by closing the channel just to implement a scorched-earth strategy.

But in most cases, the initiator will provide more funds than the non-initiator, which means the initiator receives `tx_signatures` first. When receiving those signatures, you can check witness data, and if the witnesses are much larger than what you expect, you can simply not send your `tx_signatures` and double-spend one of your inputs. If you double-spend with a transaction you intended to make anyway, you're not being penalized at all, and you effectively "preemptively closed" the channel.

Reviewer 9: If you have the peer's `tx_signatures` and haven't sent your `tx_signatures`, you can simply abort the negotiation, no double-spend necessary.  This effectively punishes the peer, since they must keep their commitment transaction around until you or *they* spend an input somewhere else.

If you have sent `tx_signatures` first and then the peer's `tx_signatures` make the feerate too low, it may be best to just broadcast the transaction and not preemptively close.  While you potentially suffer opportunity cost due to funds getting stuck in mempool, your peer (usually) suffers more since they have a larger contribution.

These incentives change though if the peer is brokering a multiparty transaction.  With a relatively small amount of their own funds (e.g, the peer is purchasing liquidity), the peer could end up with the larger contribution in each negotiation and would therefore suffer less from fund lockup than the other peers.  But other griefing attacks already exist in the LN, so maybe it's not worth worrying about at this point.

Reviewer 3: I think ultimately we will never be able to remove completely griefing attacks vectors from multiparty funding phase (or few other operations like PeerSwap) and to reduce the risks of exploitation by competing entities, we might have to introduce risk premium fees, where the premium request is function of your peer's reputation. One should be able to get coverage for the asymmetries endured due to protocol flow.

Overall, aborting the negotiation without double-spending, or letting the broadcasting flow if the damage suffers isn't the most prevalent one, sounds economically rational behaviors.

> for underpayment of fees.

Author: > Yeah, the 'double spending of an input to cancel an open' is a bit implicit here, it's probably worthwhile to call it out explicitly.

> for underpayment of fees.

Reviewer 3: This can be studied in a version 0.2 of the specification, see ongoing conversation on the mailing list on this subject: https://lists.linuxfoundation.org/pipermail/bitcoin-dev/2022-November/021142.html

        |       |                              |       |
        |       |<-(e)--  tx_signatures -------|       |
        |       |--(f)--  tx_signatures ------>|       |
        |       |                              |       |
        |       |--(7)--- channel_ready  ----->|       |
        |       |<-(8)--- channel_ready  ------|       |
        +-------+                              +-------+

        - where node A is *opener*/*initiator* and node B is
          *accepter*/*non-initiator*

### The `open_channel2` Message

This message initiates the v2 channel establishment workflow.

1. type: 64 (`open_channel2`)
2. data:
   * [`chain_hash`:`chain_hash`]
   * [`channel_id`:`temporary_channel_id`]
   * [`u32`:`funding_feerate_perkw`]
   * [`u32`:`commitment_feerate_perkw`]
   * [`u64`:`funding_satoshis`]
   * [`u64`:`dust_limit_satoshis`]
   * [`u64`:`max_htlc_value_in_flight_msat`]
   * [`u64`:`htlc_minimum_msat`]
   * [`u16`:`to_self_delay`]
   * [`u16`:`max_accepted_htlcs`]
   * [`u32`:`locktime`]
   * [`point`:`funding_pubkey`]
   * [`point`:`revocation_basepoint`]
   * [`point`:`payment_basepoint`]
   * [`point`:`delayed_payment_basepoint`]
   * [`point`:`htlc_basepoint`]

>    * [`point`:`first_per_commitment_point`]

Reviewer 8: I think I made a mistake in #25 (84 years ago), I should have also included the `second_per_commitment_point` in `open_channel` (we currently send later, in the `channel_ready`).

It doesn't change anything from a protocol point of view, but it would make the implementation easier, at least in eclair. We currently have to make a special case for the first commitments and use a placeholder value, because during normal channel operations we always have the next `commitment_point` available when we don't have a sig in-flight. Curious if other implementers ran into this?

Maybe we could take the opportunity to add it here.

Author: to check my understanding: this is a proposal to send both the `first` and `second` commitment point in `open_channel2`. The reason for this is so that the in-program data structures are correctly set with both when instantiated (with both current + next points).

Currently we send this second point in `channel_ready`, which happens at some (indeterminate) time after the channel has been committed to chain. 

Seems reasonable to me.

Might impact the `splicing` proposal, cc @ddustin ?

Author: Given that the `channel_ready` message is re-used between the v1 and v2 protocols though, I don't think we can remove it from the message. So sending it in `open_channel2` would be redundant.

Reviewer 8: > to check my understanding [...]

Your understanding is correct

I actually came to this while working on a `splicing` prototype. The natural thing to do to confirm a splice is to send a `channel_ready`, but then `channel_ready.second_per_commitment_point` is odd. This reminded me that this field shouldn't have been there in the first place.

I suppose @ddustin has run into this too? Maybe by reusing `channel_ready` isn't such a good idea if we need to introduce a new equivalent message for `splicing`.

Reviewer 2: @niftynei this should also be added to `accept_channel2`. I agree that we cannot change `channel_ready` at this point, but it's ok, we can simply ignore its `next_per_commitment_point` when using dual funding. I don't think it's worth introducing a new message just for this. Maybe once we deprecate v1 open we'll have ideas to just re-purpose that field?

Reviewer 7: > I actually came to this while working on a `splicing` prototype. The natural thing to do to confirm a splice is to send a `channel_ready`, but then `channel_ready.second_per_commitment_point` is odd. This reminded me that this field shouldn't have been there in the first place.
> 
> I suppose @ddustin has run into this too? Maybe reusing `channel_ready` isn't such a good idea if we need to introduce a new equivalent message for `splicing`.

Yup the splicing spec calls for a `splice_locked` message but doesn't yet define it. For development I've been using this message spec.

```
splice_locked(78)
+ [channel_id]channel_id
```

Reminds me we should add it explicitly into the spec. I think `channel_id` is all it needs although it may be worth thinking about the situation where we have multiple RBF competing splices and saying which one confirmed. Both nodes should really be monitoring for all of them already though 🤷

    |   |       |--(a)--- tx_init_rbf -------->|       |
    ----|       |<-(b)--- tx_ack_rbf ----------|       |
        |       |                              |       |
        |       |    <tx rbf collaboration>    |       |
        |       |                              |       |
        |       |--(c)--  commitment_signed -->|       |
        |       |<-(d)--  commitment_signed ---|       |
        |       |                              |       |
        |       |<-(e)--  tx_signatures -------|       |
        |       |--(f)--  tx_signatures ------>|       |
        |       |                              |       |
        |       |--(7)--- channel_ready  ----->|       |
        |       |<-(8)--- channel_ready  ------|       |
        +-------+                              +-------+

        - where node A is *opener*/*initiator* and node B is
          *accepter*/*non-initiator*

### The `open_channel2` Message

This message initiates the v2 channel establishment workflow.

1. type: 64 (`open_channel2`)
2. data:
   * [`chain_hash`:`chain_hash`]
   * [`channel_id`:`temporary_channel_id`]
   * [`u32`:`funding_feerate_perkw`]
   * [`u32`:`commitment_feerate_perkw`]

>    * [`u64`:`funding_satoshis`]

Reviewer 13: why do we need this funding_sathosis in the open_channel message ? We as a channelopener don't know yet how much funds the peer will contribute. Or is it here to signal in case we set it to 0, that we don't want to open a dual-funding channel but fund it by ourselves ?

Reviewer 10: This is the amount the channel opener will be contributing and doesn't include the acceptor's contribution.

as a convenience for implementations.

The sending node may require the other participant to only use confirmed inputs.
This ensures that the sending node doesn't end up paying the fees of a low
feerate unconfirmed ancestor of one of the other participant's inputs.

### The `accept_channel2` Message

This message contains information about a node and indicates its
acceptance of the new channel.

1. type: 65 (`accept_channel2`)
2. data:
    * [`channel_id`:`temporary_channel_id`]
    * [`u64`:`funding_satoshis`]
    * [`u64`:`dust_limit_satoshis`]
    * [`u64`:`max_htlc_value_in_flight_msat`]
    * [`u64`:`htlc_minimum_msat`]
    * [`u32`:`minimum_depth`]
    * [`u16`:`to_self_delay`]
    * [`u16`:`max_accepted_htlcs`]
    * [`point`:`funding_pubkey`]
    * [`point`:`revocation_basepoint`]
    * [`point`:`payment_basepoint`]
    * [`point`:`delayed_payment_basepoint`]
    * [`point`:`htlc_basepoint`]
    * [`point`:`first_per_commitment_point`]
    * [`point`:`second_per_commitment_point`]
    * [`accept_tlvs`:`tlvs`]

1. `tlv_stream`: `accept_tlvs`
2. types:
   1. type: 0 (`upfront_shutdown_script`)
   2. data:
       * [`...*byte`:`shutdown_scriptpubkey`]

>    1. type: 1 (`channel_type`)

Reviewer 10: Any chance we can make `channel_type` a requirement for V2 channels now? Would be a good opportunity.

Reviewer 2: I really don't think we should bundle features just because we want to force people to use an unrelated feature X to get feature Y. Dual funding does not depend in any way on channel types. If you want to make channel types mandatory, just make that feature mandatory on your node (which we can probably target for next releases, as every implementation has shipped support for channel types AFAIK?).

Reviewer 10: Yeah makes sense. 👍

Reviewer 2: > but if we do that we should definitely make it required here

I don't understand the reasoning here...If we make channel types mandatory in all implementations, we can then clean up the spec to make it mandatory in the spec as well (like we did with `var_onion_optin`). Why would we then need to create a dependency on it for dual funding? We'll make it mandatory in all cases, not specifically for dual funding?

Sorry if I'm missing something here, I can't see why would ever need to use feature dependencies for something else than a real logical dependency. Making a feature mandatory seems completely orthogonal to other features?

Note that `accept_channel`'s `channel_reserve_satoshi` has been omitted.
Instead, the channel reserve is fixed at 1% of the total channel balance
(`open_channel2`.`funding_satoshis` + `accept_channel2`.`funding_satoshis`)
rounded down to the nearest whole satoshi or the `dust_limit_satoshis`,
whichever is greater.

### Funding Composition

Funding composition for channel establishment v2 makes use of the
[Interactive Transaction Construction](#interactive-transaction-construction)
protocol, with the following additional caveats.

#### The `tx_add_input` Message

##### Requirements

The sending node:
  - if the receiver set `require_confirmed_inputs` in `open_channel2` or `accept_channel2`:
    - MUST NOT send a `tx_add_input` that contains an unconfirmed input

#### The `tx_add_output` Message

##### Requirements

The sending node:
  - if is the *opener*:
    - MUST send at least one `tx_add_output`,  which contains the
      channel's funding output

##### Rationale

The channel funding output must be added by the *opener*, who pays its fees.

#### The `tx_complete` Message

Upon receipt of consecutive `tx_complete`s, the receiving node:

>   - if is the *accepter*:

Author: `if is the ...` is common language across the rest of this document. will not fix.

>   - if is the *accepter*:

Reviewer 9: Actually, prior to this PR:

Ctrl+F for "if is" brings up 0 results in this document.

Ctrl+F for "if it is" brings up" 11 results in this document.

If we want consistency, we should `s/if is/if it is` across the whole PR.

### The `commitment_signed` Message

This message is exchanged by both peers. It contains the signatures for
the first commitment transaction.

Rationale and Requirements are the same as listed below,
for [`commitment_signed`](#commiting-updates-so-far-commitment_signed) with the following additions.

#### Requirements

The sending node:
  - MUST send zero HTLCs.
  - MUST remember the details of this funding transaction.

The receiving node:
  - if the message has one or more HTLCs:
    - MUST fail the negotiation
  - if it has not already transmitted its `commitment_signed`:
    - MUST send `commitment_signed`
  - Otherwise:
    - MUST send `tx_signatures` if it should sign first, as specified
      in the [`tx_signatures` requirements](#the-tx_signatures-message)

#### Rationale

The first commitment transaction has no HTLCs.

Once peers are ready to exchange commitment signatures, they must remember
the details of the funding transaction to allow resuming the signatures
exchange if a disconnection happens.

### Sharing funding signatures: `tx_signatures`

After a valid `commitment_signed` has been received
from the peer and a `commitment_signed` has been sent, a peer:

>   - MUST transmit `tx_signatures` with their signatures for the funding

Reviewer 3: I think the requirement "MUST use `SIGHASH_ALL` to commit to their peer contributed inputs outpoints" can be added, otherwise your peer can tweak the contributed inputs to break initial commitment exchanged `commitment_signed` validity.

    transaction, following the order specified in the
    [`tx_signatures` requirements](#the-tx_signatures-message)

#### Requirements

The sending node:
  - MUST verify it has received a valid commitment signature from its peer
  - MUST remember the details of this funding transaction
  - if it has NOT received a valid `commitment_signed` message:
    - MUST NOT send a `tx_signatures` message

The receiving node:
  - if the `witness` weight lowers the effective `feerate`
    below the the *opener*'s feerate for the funding transaction:

>     - SHOULD broadcast their commitment transaction, closing the channel.

Author: You're correct, the channel can still operate normally if this is violated.

The rationale describes why this is suggested: "This should disincentivize peers from underpaying fees."

Note that it's at the level of recommendation (SHOULD) not directive (MUST). I updated the ordering + some of the wording for this in an attempt to make it clearer.

>     - SHOULD broadcast their commitment transaction, closing the channel.

Reviewer 9: But does it really disincentivize the behavior?  Force closing the channel may very well hurt us more than the peer.  If we're the opener, we will likely pay significantly more fees than the peer -- more fees on the funding tx and all the fees on the commitment tx.  If we're the accepter, maybe there's a disincentive, or maybe not -- it depends on the fees paid by us on the funding tx.

I think the best course of action in this situation is to double spend productively if there's an opportunity.  Otherwise, we can just wait -- we don't have another use for those funds anyway.

And we need to have this double-spending logic in place anyway to handle griefing from withholding `tx_signatures`.  Why not use it to address this griefing vector as well?

>     - SHOULD broadcast their commitment transaction, closing the channel.

Author: Reasonably, if you don't believe that the feerate on a channel open is good enough to be mined, you're safe to broadcast the commitment signatures and double-spend your inputs to the channel.

This covers the case where the existing channel gets mined while you're waiting for the other use of those inputs to be mined/happen.

In the case that the channel does eventually open, then it'll immediately close. If it never opens, then you never spend the onchain fees.

        - where node A is *opener*/*initiator* and node B is
          *accepter*/*non-initiator*

### The `open_channel2` Message

This message initiates the v2 channel establishment workflow.

1. type: 64 (`open_channel2`)
2. data:
   * [`chain_hash`:`chain_hash`]
   * [`channel_id`:`temporary_channel_id`]
   * [`u32`:`funding_feerate_perkw`]
   * [`u32`:`commitment_feerate_perkw`]
   * [`u64`:`funding_satoshis`]
   * [`u64`:`dust_limit_satoshis`]
   * [`u64`:`max_htlc_value_in_flight_msat`]
   * [`u64`:`htlc_minimum_msat`]
   * [`u16`:`to_self_delay`]
   * [`u16`:`max_accepted_htlcs`]
   * [`u32`:`locktime`]
   * [`point`:`funding_pubkey`]
   * [`point`:`revocation_basepoint`]
   * [`point`:`payment_basepoint`]
   * [`point`:`delayed_payment_basepoint`]
   * [`point`:`htlc_basepoint`]
   * [`point`:`first_per_commitment_point`]
   * [`point`:`second_per_commitment_point`]
   * [`byte`:`channel_flags`]
   * [`opening_tlvs`:`tlvs`]

1. `tlv_stream`: `opening_tlvs`
2. types:
   1. type: 0 (`upfront_shutdown_script`)
   2. data:
       * [`...*byte`:`shutdown_scriptpubkey`]
   1. type: 1 (`channel_type`)
   2. data:
        * [`...*byte`:`type`]

>    1. type: 2 (`require_confirmed_inputs`)

Reviewer 10: If `require_confirmed_inputs` is present, would it make sense to include block hash or height for added UTXOs? It would make it practical for light clients to uphold.

Reviewer 2: Can you detail? Who would that help exactly, and how? IMO if you're a light client you simply won't require confirmed inputs from your LSP, who wouldn't accept that requirement anyway. This is rather the other way around, the node with a large volume of channel opens is the one that may require its peers to use confirmed inputs.

Reviewer 10: Just thinking if it's reasonable to assume that maybe two light clients want to open a channel with one another. One of the peers (or both) wants to ensure the other uses confirmed inputs (to avoid paying fees for low feerate ancestors) but might not have an efficient way to query the UTXO set for the counterparty's UTXO, but with a block hash/height it could query and look for inclusion.

Sorry, I probably didn't understand the original reasoning for `require_confirmed_inputs`.

Reviewer 2: > but might not have an efficient way to query the UTXO set for the counterparty's UTXO, but with a block hash/height it could query and look for inclusion.

If I'm understanding this correctly, this block hash / block height should be added to `tx_add_input`, right? You'd like every input to optionally specify the height/hash at which it was confirmed to help the receiver check that input's confirmation status?

That could be reasonable, but I don't think it makes sense in the base dual funding specification, because I'm not sure nodes would want to be forced to provide this information. In our case for example obtaining that information for our own inputs requires an additional call to `bitcoind`, which is annoying. So maybe that's something that could be added later (if we really need it for some use cases) with a dedicated feature bit (which some nodes wouldn't enable if it's costly for them to get this information).

Reviewer 10: Yeah exactly that.

I don't think it's essential to have to get this merged, but I do feel it would be useful.

Happy to defer to a future feature. 👍

the first commitment transaction.

Rationale and Requirements are the same as listed below,
for [`commitment_signed`](#commiting-updates-so-far-commitment_signed) with the following additions.

#### Requirements

The sending node:
  - MUST send zero HTLCs.

The receiving node:
  - if the message has one or more HTLCs:
    - MUST fail the negotiation
  - if it has not already transmitted its `commitment_signed`:
    - MUST send `commitment_signed`
  - Otherwise:
    - MUST send `tx_signatures` if it should sign first, as specified
      in the [`tx_signatures` requirements](#the-tx_signatures-message)

#### Rationale

The first commitment transaction has no HTLCs.

### Sharing funding signatures: `tx_signatures`

After a valid `commitment_signed` has been received
from the peer and a `commitment_signed` has been sent, a peer:
  - MUST transmit `tx_signatures` with their signatures for the funding
    transaction, following the order specified in the
    [`tx_signatures` requirements](#the-tx_signatures-message)

#### Requirements

The sending node:
  - MUST verify it has received a valid commitment signature from its peer

>   - MUST remember the details of this funding transaction

Reviewer 2: I think we should move that requirement to the sending side of `commitment_signed`. Combined with an additional mechanism to retransmit `commitment_signed` on reconnection when necessary, this gets rid of cases where one side has stored the channel state and the other hasn't (in which case we would previously need to actively double-spend the channel or wait for the other participant to actively double-spend it, which is annoying). I think that we may not even need to change anything for the reestablish logic, we can probably distinguish the case where `next_commitment_number` is `0` from the one where it is `1` to figure out whether the first `commitment_signed` needs to be retransmitted or not.

I provided more details in [this gist](https://gist.github.com/t-bast/2594ecf0314f0b91293a3f5e3bf10540), with diagrams of the various scenarios that create issues.

Reviewer 2: Here is a more detailed proposal, after doing some implementation work in eclair.

The first step is to move the storage requirement to the `commitment_signed` section: once `tx_complete` has been exchanged, we store the transaction details and send `commitment_signed`.
There is no need for an ordering on `commitment_signed`: both peers should send it as soon as `tx_complete` has been exchanged.

If a disconnection happens before `tx_signatures` has been received by both peers, we resume the signing session when reconnecting.
In that case, at least one of the peers is missing either:

- `commitment_signed` and `tx_signatures`
- only `tx_signatures` if they're already received `commitment_signed`

We introduce a new tlv to the `channel_reestablish` message to ask our peer to retransmit the missing messages.
This tlv must not be included if we've already received `commitment_signed` and `tx_signatures`.

1. `tlv_stream`: `channel_reestablish_tlvs`
2. types:
  1. type: 0 (`next_funding`)
  2. data:
    * [`sha256`:`funding_txid`]
    * [`byte`:`retransmit_commitment_signed`]

We can get rid of the existing requirement of always retransmitting `tx_signatures` now that we explicitly ask our peer to retransmit it if necessary.

A node that receives `channel_reestablish` that contains the `next_funding` tlv:

- If it doesn't have a channel for that `channel_id`:
  - MUST send an `error` and close the connection.
  - Note that the fact that the node didn't remember that channel means it never sent its `commitment_signed`, which means the other node hasn't sent its `tx_signatures` and can thus safely forget that channel when receiving the `error`.
- If `funding_txid` doesn't match one of the funding transactions:
  - MUST send `tx_abort` and resume the channel.
  - This must be for an RBF attempt where we got disconnected before storing the RBF transaction, which means we haven't sent our `commitment_signed`.
  - This `tx_abort` lets the other peer know that they can safely forget that RBF attempt (since we didn't send `commitment_signed`, they haven't sent `tx_signatures` so they're not risking anything).
- If `funding_txid` matches one of the funding transactions:
  - If `retransmit_commitment_signed` is set to `0x01`, MUST retransmit `commitment_signed` and `tx_signatures`.
  - If `retransmit_commitment_signed` is set to `0x00`, MUST retransmit `tx_signatures`.

With those changes, the following scenarios are nicely handled (variations of those scenarios are left as an exercise to the reader :D).
The corresponding scenarios will work for splicing as well, where it's crucial to correctly handle them (otherwise they would lead to a unilateral close).

## Half-forgotten channel

```text
  Alice                          Bob
    |        tx_complete          |
    |<----------------------------|
    |        tx_complete          |
    |------------->X              |
    |        commit_sig           |
    |------------->X              |
    |                             |
    |       (reconnection)        |
    |                             |
    |    channel_reestablish      |
    |---------------------------->|
    |           error             |
    |<----------------------------|
```

Alice's `channel_reestablish` contains `next_funding` with `retransmit_commitment_signed`.
Bob has forgotten that channel and sends an `error`: when receiving that `error`, Alice can simply forget that channel too.

## Half-forgotten RBF attempt

```text
  Alice                          Bob
    |        tx_complete          |
    |---------------------------->|
    |        tx_complete          |
    |<----------------------------|
    |        commit_sig           |
    |<----------------------------|
    |        commit_sig           |
    |---------------------------->|
    |       tx_signatures         |
    |---------------------------->|
    |       tx_signatures         |
    |<----------------------------|
    |        tx_init_rbf          |
    |---------------------------->|
    |        tx_ack_rbf           |
    |---------------------------->|
    |           ...               |
    |        tx_complete          |
    |<----------------------------|
    |        tx_complete          |
    |------------->X              |
    |        commit_sig           |
    |------------->X              |
    |                             |
    |       (reconnection)        |
    |                             |
    |    channel_reestablish      |
    |---------------------------->|
    |    channel_reestablish      |
    |<----------------------------|
    |          tx_abort           |
    |<----------------------------|
    |          tx_abort           |
    |---------------------------->|
```

Alice's `channel_reestablish` contains `next_funding` with `retransmit_commitment_signed`, using the `funding_txid` of the RBF transaction.
Bob has forgotten that RBF attempt on disconnection, so its `channel_reestablish` doesn't contain `next_funding`.
Alice's `next_funding.funding_txid` doesn't match anything on Bob's side, so Bob sends `tx_abort`.
Alice receives that `tx_abort`, cancels the RBF attempt (which is safe since she hasn't sent `tx_signatures`) and echoes `tx_abort`.
Both sides have correctly cleaned up that failed RBF attempt.

## Partially signed channel

```text
  Alice                          Bob
    |        tx_complete          |
    |---------------------------->|
    |        tx_complete          |
    |<----------------------------|
    |        commit_sig           |
    |<----------------------------|
    |        commit_sig           |
    |--------------->X            |
    |       tx_signatures         |
    |--------------->X            |
    |                             |
    |       (reconnection)        |
    |                             |
    |    channel_reestablish      |
    |---------------------------->|
    |    channel_reestablish      |
    |<----------------------------|
    |        commit_sig           |
    |---------------------------->|
    |       tx_signatures         |
    |---------------------------->|
    |       tx_signatures         |
    |<----------------------------|
```

In this example, Alice was supposed to send `tx_signatures` first according to the ordering requirements.
Alice's `channel_reestablish` contains `next_funding` without `retransmit_commitment_signed`.
Bob's `channel_reestablish` contains `next_funding` with `retransmit_commitment_signed`.
Alice retransmits `commitment_signed` and `tx_signatures`, and Bob sends his `tx_signatures` once he has received Alice's.
The channel is now fully functional.

## Partially signed RBF attempt

```text
  Alice                          Bob
    |        tx_complete          |
    |---------------------------->|
    |        tx_complete          |
    |<----------------------------|
    |        commit_sig           |
    |<----------------------------|
    |        commit_sig           |
    |---------------------------->|
    |       tx_signatures         |
    |---------------------------->|
    |       tx_signatures         |
    |<----------------------------|
    |        tx_init_rbf          |
    |---------------------------->|
    |        tx_ack_rbf           |
    |---------------------------->|
    |           ...               |
    |        tx_complete          |
    |<----------------------------|
    |        tx_complete          |
    |---------------------------->|
    |        commit_sig           |
    |<----------------------------|
    |        commit_sig           |
    |--------------->X            |
    |       tx_signatures         |
    |--------------->X            |
    |                             |
    |       (reconnection)        |
    |                             |
    |    channel_reestablish      |
    |---------------------------->|
    |    channel_reestablish      |
    |<----------------------------|
    |        commit_sig           |
    |---------------------------->|
    |       tx_signatures         |
    |---------------------------->|
    |       tx_signatures         |
    |<----------------------------|
```

In this example, Alice was supposed to send `tx_signatures` first according to the ordering requirements.
Alice's `channel_reestablish` contains `next_funding` without `retransmit_commitment_signed`, using the `funding_txid` of the RBF attempt.
Bob's `channel_reestablish` contains `next_funding` with `retransmit_commitment_signed`, using the `funding_txid` of the RBF attempt.
Alice retransmits `commitment_signed` and `tx_signatures`, and Bob sends his `tx_signatures` once he has received Alice's.
This completes the RBF attempt.

@niftynei @ddustin @rustyrussell I believe that this is also a better way of handling disconnection/reestablish for splicing, compared to incrementing the `commitment_index` like you currently proposed (which requires re-signing all active commitments and revoking the previous one for no added benefits: see https://github.com/lightning/bolts/pull/863#discussion_r1135122533 for the corresponding splice discussion).

Reviewer 2: We can even simplify this by getting rid of the `retransmit_commitment_signed` field, and whenever the `next_funding` tlv is present, always retransmit both `commitment_signed` and `tx_signatures`. This has the added benefit that we don't need to persist the remote `commitment_signed` when we receive it, we can wait for the `tx_signatures` exchange before updating our persisted state.

Reviewer 2: If you want to have a look at the implementation diff for that change, you can find it [here](https://github.com/ACINQ/eclair/pull/2614) for eclair.

Reviewer 9: > Note that the fact that the node didn't remember that channel means it never sent its commitment_signed, which means the other node hasn't sent its tx_signatures and can thus safely forget that channel when receiving the error.

...

> This tx_abort lets the other peer know that they can safely forget that RBF attempt (since we didn't send commitment_signed, they haven't sent tx_signatures so they're not risking anything).

One thing to be careful about is not assuming the peer is honest.  We don't want to forget a channel or RBF attempt *only* because the peer responded with `error` or `tx_abort`.  We also need to verify from our own records that we haven't sent `tx_signatures`.

Reviewer 2: Yes definitely: once you've sent `tx_signatures`, you need to remember the channel until it has been spent or double-spent, regardless of what your peer tells you.

Reviewer 2: Since splicing is another user of `interactive-tx`, reviewers may be interested in our [splicing proposal](https://gist.github.com/t-bast/1ac31f4e27734a10c5b9847d06db8d86), in which we also use these `channel_reestablish` changes (because without it, a disconnection at the wrong time would lead to a force-close, which isn't acceptable).

Reviewer 2: Here is a patch including the changes proposed: https://github.com/niftynei/lightning-rfc/pull/13

Reviewer 7: ACK.

Love that we don’t need to store commitment signed state in the DB by just resending it.

A few thoughts: you should probably enter the splice commitment process when *sending* next_funding tlv, since signing order could be you first.

Following that you should *not* enter a second commitment process when receiving the next_funding tlv, so you don’t end up doing it twice.

Easy enough reason about thought, not sure it requires explicitly mentioning in the spec.

“next_funding” is a little obscure of a name to me. Something like “retry_splice” would be a much clearer name to me, unless this is being used for other purposes as well.

Lastly we may want to be more explicit about when to consider the splice reestablish-able. I would propose the specific moment we consider it that is when we receive tx_complete.

Reviewer 2: > Easy enough reason about though, not sure it requires explicitly mentioning in the spec.

Agreed, I don't  think this is worth mentioning. At least it's clearly not worth mentioning now, because for dual funding it simply doesn't apply. We can see if it makes sense to add in the splice PR when we make more progress on the splice spec PR.

> “next_funding” is a little obscure of a name to me. Something like “retry_splice” would be a much clearer name to me, unless this is being used for other purposes as well.

But this is used for dual funding as well, so it definitely cannot be named `retry_splice`! It is useful for any `interactive-tx` session, which will also very likely encompass mutual closing in the future, so we want something generic enough. This is for the next transaction spending the current funding output, and for dual funding and splicing it is a funding transaction: this is why I think `next_funding_tx` makes sense. But if we start using this for closing txs, I guess we can only do `next_interactive_tx` to cover all cases, but that's something we can change when we specify interactive-tx-based closing.

> Lastly we may want to be more explicit about when to consider the splice reestablish-able. I would propose the specific moment we consider it that is when we receive tx_complete.

Again, this is the dual funding PR, we shouldn't mention splice, and the use-case it currently solves is unrelated to splice. IMO we should consider the interactive-tx reestablish-able as soon as we send our `commit_sig`, which is what is currently specified.

the details of the funding transaction to allow resuming the signatures
exchange if a disconnection happens.

### Sharing funding signatures: `tx_signatures`

After a valid `commitment_signed` has been received
from the peer and a `commitment_signed` has been sent, a peer:
  - MUST transmit `tx_signatures` with their signatures for the funding
    transaction, following the order specified in the
    [`tx_signatures` requirements](#the-tx_signatures-message)

#### Requirements

The sending node:
  - MUST verify it has received a valid commitment signature from its peer
  - MUST remember the details of this funding transaction
  - if it has NOT received a valid `commitment_signed` message:
    - MUST NOT send a `tx_signatures` message

The receiving node:
  - if has already sent or received a `channel_ready` message for this
    channel:
    - MUST ignore this message
  - if the `witness` weight lowers the effective `feerate`
    below the *opener*'s feerate for the funding transaction and the effective
	`feerate` is determined by the receiving node to be insufficient for
	getting the transaction confirmed in a timely manner:

>     - SHOULD broadcast their commitment transaction, closing the channel

Reviewer 9: As discussed on the other comment thread, I think we should remove the recommendation to broadcast the commitment, since that may actually punish ourselves more than the fee-underpaying peer.  And broadcasting the commitment only matters if the funding tx actually confirms, in which case we now have a working channel.  It seems a little silly to force close a working channel out of spite.

The existing recommendation to productively double-spend if the fee is underpaid addresses the potential griefing vector.  If the funding tx confirms, then either no griefing happened in practice or it ended before we could productively use those funds anyway.  No harm, no foul, so let's not make things worse by force closing?

```suggestion
```

Also, (nit) the indentation is off in the recent changes here.

Author: > broadcasting the commitment only matters if the funding tx actually confirms, in which case we now have a working channel. It seems a little silly to force close a working channel out of spite.

That's not entirely correct: I'd expect the commitment transaction to sit in the mempool along with the unmined opening transaction. In the case that the channel does open, it will be closed by the commitment transaction. I don't see any downsides to broadcasting a close in the case that you're attempting to re-use inputs in the funding elsewhere. You've already so much as considered this channel dead on arrival, there's no harm in publishing a transaction you're not anticipating will be mined anyway.

Also note, the force close is recommended but not mandatory behavior. There's a real benefit to the peer knowing that if they underpay the channel likely won't open.

Reviewer 9: I feel like we're talking past each other.  I'll try to respond directly to each of your comments below.  Please let me know where I'm incorrect or missing something.

> > broadcasting the commitment only matters if the funding tx actually confirms, in which case we now have a working channel. It seems a little silly to force close a working channel out of spite.
> 
> That's not entirely correct: I'd expect the commitment transaction to sit in the mempool along with the unmined opening transaction. In the case that the channel does open, it will be closed by the commitment transaction.

I think it's exactly correct.  As you've just said, the commitment tx can only confirm if the channel opens.  If the channel never opens, and therefore the commitment tx never confirms, does the fact that you broadcasted commitment tx really matter?

And if the funding tx does confirm, any griefing is over, so there's little benefit to force closing at that point.

> I don't see any downsides to broadcasting a close in the case that you're attempting to re-use inputs in the funding elsewhere. You've already so much as considered this channel dead on arrival, there's no harm in publishing a transaction you're not anticipating will be mined anyway.

Here's 3 downsides to broadcasting the commitment tx:

1. Loss of funds -- If you're the opener and the commitment tx confirms, you probably pay more fees (funding fees + commitment fees) than if you had double spent (funding fees only).  This is especially true if your node does CPFP on the anchor output of the commitment, which I suspect most implementations currently do by default.
2. Loss of opportunity -- If the funding tx confirms, you could have actually used that channel.  But, by broadcasting the commitment tx, you've preemptively force-closed the channel instead.
3. More complexity -- Teaching your node to do 2 things (broadcast commitment + double spend) is more complex than teaching it to do 1 thing (double spend).

> Also note, the force close is recommended but not mandatory behavior.

Yes, but the recommendation in the spec is important.  We should not expect future implementers to find this PR and read our current discussion about tradeoffs before implementing.

> There's a real benefit to the peer knowing that if they underpay the channel likely won't open.

Yes, and I think the recommendation to double spend achieves this well enough on its own.  The peer knows that if they underpay, their counterparty will double spend at their earliest convenience, therefore the channel likely won't open.

Author: > I feel like we're talking past each other. 

From where I'm sitting, it appears we fundamentally disagree about what the proper behavior should be. 

Ultimately I think it's crucial to remember that underpaying fees is a bug. An implementation can avoid the undesirable outcomes of this situation by paying the proper amount for any witness data they add to a transaction.

Having a node force close for underpayment makes it much easier to spot + fix bugs in implementations, which is ultimately the goal imo and hence the spec recommendation.

Author: I disagree about this being a practical griefing vector. It's costly for a peer to execute: they would have had to include inputs that they're paying for in order to trigger this, and they would have had to paid for everything *except* the witness data at the agreed fee rate for the transaction. Why would they do all that just to 'maliciously' underpay on witness data at the last moment? Seems like behavior that the attempt of should be discouraged.

Also, the risk of your peer sending you a channel error or asking you to close the channel exist in other parts of the protocol, this isn't a new 'griefing' vector by any means.

Reviewer 2: > I disagree about this being a practical griefing vector. It's costly for a peer to execute: they would have had to include inputs that they're paying for in order to trigger this, and they would have had to paid for everything except the witness data at the agreed fee rate for the transaction.

That's a good point, it's not completely free for the attacker either, they also have to pay on-chain fees.
Since those requirements are a `SHOULD`, that works for me then, I don't mind having it and letting implementations decide whether they want to broadcast their commit or not!

Reviewer 9: > And I agree that the severity of this griefing vector is low relative to the simpler and more effective griefing vectors in the interactive-tx protocol -- e.g., the funding tx can be pinned in the mempool, requiring the aggrieved party to burn a ridiculous amount of fees to double-spend the funding tx.

Hmm... I think I just convinced myself there's a scenario where broadcasting the commitment tx makes *more* sense than trying to double spend.

If the griefer inflates their witness such that the funding tx will never confirm and also pins the funding tx with minimum fee-rate descendants, the victim's fee cost to double spend the funding tx can become ridiculous.  But if the victim does CPFP on the funding tx instead, they can get it to confirm with relatively low fees.  The CPFP can be done via commitment tx + anchor, or by spending a change output in the funding tx.

Perhaps we should mention this scenario and mitigation in the spec?

Reviewer 9: CC: @ariard

Reviewer 2: Pinning, it's always pinning... :sweat_smile: 

> The CPFP can be done via commitment tx + anchor, or by spending a change output in the funding tx.

True, but it wouldn't always work though: if the funding tx is evicted from the mempool, you won't be able to CPFP it with the commitment + anchor (even package relay won't work, because it only handles two generations of txs, not three).

But that's ok! Because if the funding tx is evicted from the mempool, then it becomes free to double-spend (the pinning is gone). So your strategy in that case should be to keep trying both options in parallel, and one of them should eventually work.

Reviewer 5: > The CPFP can be done via commitment tx + anchor, or by spending a change output in the funding tx.

Long(er) chains of transactions are less likely to propagate in general properly, and future opt-in like V3 policies may disallow this entirely, so all else equal stick to shorter CPFPs.

Reviewer 9: I'm afraid there's no good defense against pinning the funding tx.  The griefer can make RBF prohibitively expensive *and* also block CPFP by saturating the package size limit.

For commitment txs, we have the CSV 1 trick and anchors with carve out.  But we don't have a great way to enforce CSV 1 on funding tx outputs -- and probably we don't want to.

For commitment txs, pinning can cause loss of funds due to HTLC timeout.  But for funding txs the problem is griefing (fund lockup) rather than loss of funds, and to carry out the griefing the attacker must be willing to pay significant fees eventually.  Perhaps we can just accept that if the attacker is willing to burn enough fees, they can lock up our funds for a long time.

We actually [previously discussed](https://github.com/niftynei/lightning-rfc/pull/12#discussion_r1102962560) this class of griefing vectors against the interactive-tx protocol.  Back then, we agreed that "just live with it" may be the best course for now.

--------------------------------

My suggestion for the spec is to recommend both double spending and CPFP as methods to unlock funds or get the funding tx to confirm, but to note that neither is guaranteed to work if the peer is actively trying to grief us.

Reviewer 9: Revisiting this...

For simplicity, I think we can leave discussion of pinning out of the spec for now.  Since we have no great defense against it, there's not much to discuss.

But I still think the current suggested handling of witness inflation is bad -- by broadcasting the commitment tx, we force our future double spend attempts to pay ~2x the fee (RBF rule #3).  So if the peer is just buggy (not malicious), this is an expensive way to recover our funds.

Reviewer 2: I agree, those are subtle issues that aren't obvious to a spec reader today. How would you address that? By adding more details in the rationale section? Can you propose a patch that you think makes it better?

Reviewer 9: I suggested a change on the first comment in this thread.  I think removing that single line is sufficient.

If you think there's more we should include, I can attempt a larger patch.

Reviewer 2: > My suggestion for the spec is to recommend both double spending and CPFP as methods to unlock funds or get the funding tx to confirm, but to note that neither is guaranteed to work if the peer is actively trying to grief us.

I think that made the most sense instead of simply removing `SHOULD broadcast their commitment transaction`, but that requires explaining that it depends on the situation. You don't think we should suggest both options anymore?

Reviewer 9: I think there's two potential situations -- buggy implementation and malicious liquidity griefing.

#### Buggy implementation

Double spending opportunistically is the best course.  Broadcasting the commitment transaction only increases our costs of double spending.

#### Liquidity griefing

We have no foolproof defense against this currently, and any RBF or CPFP recommendation can be defeated by pinning the inflated funding transaction.

Unless we are going to recommend one of the solutions [discussed](https://delvingbitcoin.org/t/liquidity-griefing-in-multi-party-transaction-protocols/264#tldr-1) on Discourse (i.e. initiator signs first, require p2wpkh inputs, or CSV 1 on outputs), our best course of action is probably to just unlock the UTXOs and hope our double spends somehow get mined, perhaps because the pinned tx is evicted from mempools.

Reviewer 2: That makes sense to me, on the eclair side I'm planning to just unlock the utxos in that case. I'm fine with removing the `SHOULD` then if we convince @niftynei !
